{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process GREIN Human Data\n",
    "\n",
    "Retrieve the downloaded expression data, update gene identifiers to entrez, and curate sample IDs. The script will also identify a balanced hold-out test set to compare projection performance into learned latent spaces across algorithms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import random\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(1234)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Phenotype Information: skipping until phenotype data found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path = os.path.join('download', 'TARGET_phenotype.gz')\n",
    "# pheno_df = pd.read_table(path)\n",
    "\n",
    "# print(pheno_df.shape)\n",
    "# pheno_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Probe Mapping Info \n",
    "(where chromosomes start and end, saved in a file in downloads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path = os.path.join('download', 'gencode.v23.annotation.gene.probemap')\n",
    "# probe_map_df = pd.read_table(path)\n",
    "\n",
    "# # Inner merge gene df to get ensembl to entrez mapping\n",
    "# probe_map_df = probe_map_df.merge(gene_df, how='inner', left_on='gene', right_on='symbol')\n",
    "\n",
    "# # Mapping to rename gene expression index\n",
    "# ensembl_to_entrez = dict(zip(probe_map_df.id, probe_map_df.entrez_gene_id))\n",
    "\n",
    "# print(probe_map_df.shape)\n",
    "# probe_map_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Gene Expression Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(27990, 57)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>GSM2667747</th>\n",
       "      <th>GSM2667748</th>\n",
       "      <th>GSM2667749</th>\n",
       "      <th>GSM2667750</th>\n",
       "      <th>GSM2667751</th>\n",
       "      <th>GSM2667752</th>\n",
       "      <th>GSM2667753</th>\n",
       "      <th>GSM2667754</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2668086</th>\n",
       "      <th>GSM2668087</th>\n",
       "      <th>GSM2668088</th>\n",
       "      <th>GSM2668089</th>\n",
       "      <th>GSM2671001</th>\n",
       "      <th>GSM2671002</th>\n",
       "      <th>GSM2671003</th>\n",
       "      <th>GSM2671004</th>\n",
       "      <th>GSM2671005</th>\n",
       "      <th>GSM2671006</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1766.0677</td>\n",
       "      <td>420.8206</td>\n",
       "      <td>300.8009</td>\n",
       "      <td>3142.8439</td>\n",
       "      <td>2207.8392</td>\n",
       "      <td>4367.5056</td>\n",
       "      <td>228.5759</td>\n",
       "      <td>650.3287</td>\n",
       "      <td>...</td>\n",
       "      <td>254.7605</td>\n",
       "      <td>143.2263</td>\n",
       "      <td>174.3319</td>\n",
       "      <td>202.3253</td>\n",
       "      <td>1053.8788</td>\n",
       "      <td>2405.3001</td>\n",
       "      <td>940.5770</td>\n",
       "      <td>973.7475</td>\n",
       "      <td>760.0100</td>\n",
       "      <td>1472.2712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>43.9280</td>\n",
       "      <td>18.0091</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>115.8917</td>\n",
       "      <td>63.5936</td>\n",
       "      <td>39.7803</td>\n",
       "      <td>11.7588</td>\n",
       "      <td>1.0153</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3.9978</td>\n",
       "      <td>11.0025</td>\n",
       "      <td>4.9979</td>\n",
       "      <td>8.0054</td>\n",
       "      <td>3.9978</td>\n",
       "      <td>11.0011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>1097.7855</td>\n",
       "      <td>367.2333</td>\n",
       "      <td>316.5226</td>\n",
       "      <td>3895.0588</td>\n",
       "      <td>1536.1375</td>\n",
       "      <td>1084.5276</td>\n",
       "      <td>127.7205</td>\n",
       "      <td>245.1014</td>\n",
       "      <td>...</td>\n",
       "      <td>613.1214</td>\n",
       "      <td>374.0965</td>\n",
       "      <td>302.6622</td>\n",
       "      <td>454.0570</td>\n",
       "      <td>1227.1084</td>\n",
       "      <td>2980.0756</td>\n",
       "      <td>1045.8061</td>\n",
       "      <td>1098.7163</td>\n",
       "      <td>759.7729</td>\n",
       "      <td>1398.8077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>601.4702</td>\n",
       "      <td>270.9239</td>\n",
       "      <td>163.6869</td>\n",
       "      <td>1177.2241</td>\n",
       "      <td>873.9074</td>\n",
       "      <td>845.1423</td>\n",
       "      <td>229.8548</td>\n",
       "      <td>134.4694</td>\n",
       "      <td>...</td>\n",
       "      <td>196.5995</td>\n",
       "      <td>185.3106</td>\n",
       "      <td>260.9485</td>\n",
       "      <td>183.7584</td>\n",
       "      <td>586.7993</td>\n",
       "      <td>1504.0115</td>\n",
       "      <td>619.3764</td>\n",
       "      <td>661.9514</td>\n",
       "      <td>469.5312</td>\n",
       "      <td>1113.1412</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows Ã— 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene gene_symbol  GSM2667747  GSM2667748  GSM2667749  \\\n",
       "0  ENSG00000000003      TSPAN6   1766.0677    420.8206    300.8009   \n",
       "1  ENSG00000000005        TNMD     43.9280     18.0091      0.0000   \n",
       "2  ENSG00000000419        DPM1   1097.7855    367.2333    316.5226   \n",
       "3  ENSG00000000457       SCYL3    601.4702    270.9239    163.6869   \n",
       "\n",
       "   GSM2667750  GSM2667751  GSM2667752  GSM2667753  GSM2667754  ...  \\\n",
       "0   3142.8439   2207.8392   4367.5056    228.5759    650.3287  ...   \n",
       "1    115.8917     63.5936     39.7803     11.7588      1.0153  ...   \n",
       "2   3895.0588   1536.1375   1084.5276    127.7205    245.1014  ...   \n",
       "3   1177.2241    873.9074    845.1423    229.8548    134.4694  ...   \n",
       "\n",
       "   GSM2668086  GSM2668087  GSM2668088  GSM2668089  GSM2671001  GSM2671002  \\\n",
       "0    254.7605    143.2263    174.3319    202.3253   1053.8788   2405.3001   \n",
       "1      0.0000      0.0000      0.0000      0.0000      3.9978     11.0025   \n",
       "2    613.1214    374.0965    302.6622    454.0570   1227.1084   2980.0756   \n",
       "3    196.5995    185.3106    260.9485    183.7584    586.7993   1504.0115   \n",
       "\n",
       "   GSM2671003  GSM2671004  GSM2671005  GSM2671006  \n",
       "0    940.5770    973.7475    760.0100   1472.2712  \n",
       "1      4.9979      8.0054      3.9978     11.0011  \n",
       "2   1045.8061   1098.7163    759.7729   1398.8077  \n",
       "3    619.3764    661.9514    469.5312   1113.1412  \n",
       "\n",
       "[4 rows x 57 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reading in rat gene expression data\n",
    "\n",
    "file = os.path.join('download', 'grein_count_matrix_human.pkl')\n",
    "expr_df = pd.read_pickle(file)\n",
    "\n",
    "print(expr_df.shape)\n",
    "expr_df.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean up x and y from expr_df\n",
    "\n",
    "# create a list of cleaned columns\n",
    "cleaned_column = ['gene', 'gene_symbol']\n",
    "for column in expr_df.drop(['gene', 'gene_symbol'], axis = 1).columns:\n",
    "    cleaned_column.append(column.split(\"_\")[0])\n",
    "\n",
    "# set colums to cleaned column\n",
    "expr_df.columns = cleaned_column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>GSM2667747</th>\n",
       "      <th>GSM2667748</th>\n",
       "      <th>GSM2667749</th>\n",
       "      <th>GSM2667750</th>\n",
       "      <th>GSM2667751</th>\n",
       "      <th>GSM2667752</th>\n",
       "      <th>GSM2667753</th>\n",
       "      <th>GSM2667754</th>\n",
       "      <th>GSM2667755</th>\n",
       "      <th>GSM2667756</th>\n",
       "      <th>GSM2667757</th>\n",
       "      <th>GSM2667758</th>\n",
       "      <th>GSM2667759</th>\n",
       "      <th>GSM2667760</th>\n",
       "      <th>GSM2667761</th>\n",
       "      <th>GSM2667762</th>\n",
       "      <th>GSM2667763</th>\n",
       "      <th>GSM2667764</th>\n",
       "      <th>GSM2667765</th>\n",
       "      <th>GSM2667766</th>\n",
       "      <th>GSM2667767</th>\n",
       "      <th>GSM2667768</th>\n",
       "      <th>GSM2667769</th>\n",
       "      <th>GSM2667770</th>\n",
       "      <th>GSM2667771</th>\n",
       "      <th>GSM2667772</th>\n",
       "      <th>GSM2667773</th>\n",
       "      <th>GSM2667774</th>\n",
       "      <th>GSM2667775</th>\n",
       "      <th>GSM2667776</th>\n",
       "      <th>GSM2667777</th>\n",
       "      <th>GSM2667778</th>\n",
       "      <th>GSM2667779</th>\n",
       "      <th>GSM2667780</th>\n",
       "      <th>GSM2667781</th>\n",
       "      <th>GSM2667782</th>\n",
       "      <th>GSM2667783</th>\n",
       "      <th>GSM2667784</th>\n",
       "      <th>GSM2667785</th>\n",
       "      <th>GSM2667786</th>\n",
       "      <th>GSM2668081</th>\n",
       "      <th>GSM2668082</th>\n",
       "      <th>GSM2668083</th>\n",
       "      <th>GSM2668084</th>\n",
       "      <th>GSM2668085</th>\n",
       "      <th>GSM2668086</th>\n",
       "      <th>GSM2668087</th>\n",
       "      <th>GSM2668088</th>\n",
       "      <th>GSM2668089</th>\n",
       "      <th>GSM2671001</th>\n",
       "      <th>GSM2671002</th>\n",
       "      <th>GSM2671003</th>\n",
       "      <th>GSM2671004</th>\n",
       "      <th>GSM2671005</th>\n",
       "      <th>GSM2671006</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1766.0677</td>\n",
       "      <td>420.8206</td>\n",
       "      <td>300.8009</td>\n",
       "      <td>3142.8439</td>\n",
       "      <td>2207.8392</td>\n",
       "      <td>4367.5056</td>\n",
       "      <td>228.5759</td>\n",
       "      <td>650.3287</td>\n",
       "      <td>4880.0185</td>\n",
       "      <td>8629.5404</td>\n",
       "      <td>2951.1930</td>\n",
       "      <td>370.2859</td>\n",
       "      <td>321.4122</td>\n",
       "      <td>3721.4916</td>\n",
       "      <td>3499.9075</td>\n",
       "      <td>2462.5672</td>\n",
       "      <td>300.7646</td>\n",
       "      <td>370.7972</td>\n",
       "      <td>3470.2847</td>\n",
       "      <td>3503.2877</td>\n",
       "      <td>3362.5934</td>\n",
       "      <td>226.8240</td>\n",
       "      <td>131.7482</td>\n",
       "      <td>4486.6134</td>\n",
       "      <td>7225.4738</td>\n",
       "      <td>3185.4922</td>\n",
       "      <td>531.7498</td>\n",
       "      <td>796.0883</td>\n",
       "      <td>3175.4423</td>\n",
       "      <td>2064.0788</td>\n",
       "      <td>1267.6230</td>\n",
       "      <td>1845.9171</td>\n",
       "      <td>409.7910</td>\n",
       "      <td>2540.8740</td>\n",
       "      <td>2906.8171</td>\n",
       "      <td>2829.6631</td>\n",
       "      <td>1920.6238</td>\n",
       "      <td>1165.6860</td>\n",
       "      <td>1874.8260</td>\n",
       "      <td>2539.2382</td>\n",
       "      <td>265.0806</td>\n",
       "      <td>300.3062</td>\n",
       "      <td>331.6881</td>\n",
       "      <td>183.7755</td>\n",
       "      <td>278.5022</td>\n",
       "      <td>254.7605</td>\n",
       "      <td>143.2263</td>\n",
       "      <td>174.3319</td>\n",
       "      <td>202.3253</td>\n",
       "      <td>1053.8788</td>\n",
       "      <td>2405.3001</td>\n",
       "      <td>940.5770</td>\n",
       "      <td>973.7475</td>\n",
       "      <td>760.0100</td>\n",
       "      <td>1472.2712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>43.9280</td>\n",
       "      <td>18.0091</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>115.8917</td>\n",
       "      <td>63.5936</td>\n",
       "      <td>39.7803</td>\n",
       "      <td>11.7588</td>\n",
       "      <td>1.0153</td>\n",
       "      <td>67.1228</td>\n",
       "      <td>61.4995</td>\n",
       "      <td>54.1517</td>\n",
       "      <td>11.8111</td>\n",
       "      <td>5.7130</td>\n",
       "      <td>79.6294</td>\n",
       "      <td>79.6162</td>\n",
       "      <td>69.2199</td>\n",
       "      <td>13.8240</td>\n",
       "      <td>5.9650</td>\n",
       "      <td>138.7558</td>\n",
       "      <td>84.4619</td>\n",
       "      <td>181.6344</td>\n",
       "      <td>46.6972</td>\n",
       "      <td>12.8469</td>\n",
       "      <td>168.0059</td>\n",
       "      <td>403.5948</td>\n",
       "      <td>58.7409</td>\n",
       "      <td>18.8835</td>\n",
       "      <td>12.0940</td>\n",
       "      <td>67.2416</td>\n",
       "      <td>10.9099</td>\n",
       "      <td>133.5243</td>\n",
       "      <td>147.7444</td>\n",
       "      <td>15.2011</td>\n",
       "      <td>202.6725</td>\n",
       "      <td>343.3962</td>\n",
       "      <td>676.6289</td>\n",
       "      <td>376.7828</td>\n",
       "      <td>141.2399</td>\n",
       "      <td>380.3029</td>\n",
       "      <td>662.0918</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3.9978</td>\n",
       "      <td>11.0025</td>\n",
       "      <td>4.9979</td>\n",
       "      <td>8.0054</td>\n",
       "      <td>3.9978</td>\n",
       "      <td>11.0011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>1097.7855</td>\n",
       "      <td>367.2333</td>\n",
       "      <td>316.5226</td>\n",
       "      <td>3895.0588</td>\n",
       "      <td>1536.1375</td>\n",
       "      <td>1084.5276</td>\n",
       "      <td>127.7205</td>\n",
       "      <td>245.1014</td>\n",
       "      <td>2608.6175</td>\n",
       "      <td>3797.5758</td>\n",
       "      <td>855.0836</td>\n",
       "      <td>124.4127</td>\n",
       "      <td>220.1181</td>\n",
       "      <td>2932.3449</td>\n",
       "      <td>1795.7509</td>\n",
       "      <td>1499.5849</td>\n",
       "      <td>264.1154</td>\n",
       "      <td>536.7919</td>\n",
       "      <td>3716.6839</td>\n",
       "      <td>1998.1502</td>\n",
       "      <td>1210.7017</td>\n",
       "      <td>112.5526</td>\n",
       "      <td>150.2928</td>\n",
       "      <td>3655.9428</td>\n",
       "      <td>2458.6382</td>\n",
       "      <td>834.2019</td>\n",
       "      <td>135.4603</td>\n",
       "      <td>258.5075</td>\n",
       "      <td>2264.3384</td>\n",
       "      <td>1241.8778</td>\n",
       "      <td>217.2588</td>\n",
       "      <td>623.3691</td>\n",
       "      <td>165.4709</td>\n",
       "      <td>1166.0064</td>\n",
       "      <td>1281.7182</td>\n",
       "      <td>952.9915</td>\n",
       "      <td>635.3723</td>\n",
       "      <td>879.0727</td>\n",
       "      <td>1695.9609</td>\n",
       "      <td>1736.9615</td>\n",
       "      <td>537.7681</td>\n",
       "      <td>461.4597</td>\n",
       "      <td>436.1302</td>\n",
       "      <td>416.2508</td>\n",
       "      <td>568.8483</td>\n",
       "      <td>613.1214</td>\n",
       "      <td>374.0965</td>\n",
       "      <td>302.6622</td>\n",
       "      <td>454.0570</td>\n",
       "      <td>1227.1084</td>\n",
       "      <td>2980.0756</td>\n",
       "      <td>1045.8061</td>\n",
       "      <td>1098.7163</td>\n",
       "      <td>759.7729</td>\n",
       "      <td>1398.8077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>601.4702</td>\n",
       "      <td>270.9239</td>\n",
       "      <td>163.6869</td>\n",
       "      <td>1177.2241</td>\n",
       "      <td>873.9074</td>\n",
       "      <td>845.1423</td>\n",
       "      <td>229.8548</td>\n",
       "      <td>134.4694</td>\n",
       "      <td>676.4072</td>\n",
       "      <td>1177.2415</td>\n",
       "      <td>1155.1429</td>\n",
       "      <td>232.3545</td>\n",
       "      <td>115.9401</td>\n",
       "      <td>1458.9760</td>\n",
       "      <td>1182.6409</td>\n",
       "      <td>907.3663</td>\n",
       "      <td>335.9731</td>\n",
       "      <td>153.7599</td>\n",
       "      <td>1083.9457</td>\n",
       "      <td>711.6130</td>\n",
       "      <td>1259.2643</td>\n",
       "      <td>275.6282</td>\n",
       "      <td>30.2989</td>\n",
       "      <td>1144.3081</td>\n",
       "      <td>1374.0298</td>\n",
       "      <td>1118.6175</td>\n",
       "      <td>325.3986</td>\n",
       "      <td>202.9192</td>\n",
       "      <td>1189.7186</td>\n",
       "      <td>881.3931</td>\n",
       "      <td>124.4839</td>\n",
       "      <td>948.1346</td>\n",
       "      <td>57.5744</td>\n",
       "      <td>547.4528</td>\n",
       "      <td>439.9557</td>\n",
       "      <td>1454.0821</td>\n",
       "      <td>1793.4598</td>\n",
       "      <td>232.8164</td>\n",
       "      <td>773.3686</td>\n",
       "      <td>939.7315</td>\n",
       "      <td>157.2023</td>\n",
       "      <td>151.5847</td>\n",
       "      <td>196.0139</td>\n",
       "      <td>142.7542</td>\n",
       "      <td>170.0119</td>\n",
       "      <td>196.5995</td>\n",
       "      <td>185.3106</td>\n",
       "      <td>260.9485</td>\n",
       "      <td>183.7584</td>\n",
       "      <td>586.7993</td>\n",
       "      <td>1504.0115</td>\n",
       "      <td>619.3764</td>\n",
       "      <td>661.9514</td>\n",
       "      <td>469.5312</td>\n",
       "      <td>1113.1412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>1040.8861</td>\n",
       "      <td>349.1418</td>\n",
       "      <td>298.1800</td>\n",
       "      <td>1967.3975</td>\n",
       "      <td>2115.2425</td>\n",
       "      <td>980.9359</td>\n",
       "      <td>253.4457</td>\n",
       "      <td>145.7038</td>\n",
       "      <td>1024.4449</td>\n",
       "      <td>1294.0961</td>\n",
       "      <td>599.8307</td>\n",
       "      <td>117.7714</td>\n",
       "      <td>67.3678</td>\n",
       "      <td>785.8567</td>\n",
       "      <td>466.6687</td>\n",
       "      <td>1315.1359</td>\n",
       "      <td>300.5638</td>\n",
       "      <td>246.0739</td>\n",
       "      <td>1469.0028</td>\n",
       "      <td>1906.7889</td>\n",
       "      <td>1471.1647</td>\n",
       "      <td>229.3009</td>\n",
       "      <td>67.8212</td>\n",
       "      <td>1176.1864</td>\n",
       "      <td>1307.8388</td>\n",
       "      <td>678.5934</td>\n",
       "      <td>177.1994</td>\n",
       "      <td>92.6689</td>\n",
       "      <td>560.3419</td>\n",
       "      <td>386.5186</td>\n",
       "      <td>92.7886</td>\n",
       "      <td>354.3580</td>\n",
       "      <td>49.4149</td>\n",
       "      <td>211.9013</td>\n",
       "      <td>110.9184</td>\n",
       "      <td>673.9635</td>\n",
       "      <td>749.7606</td>\n",
       "      <td>86.1525</td>\n",
       "      <td>266.8906</td>\n",
       "      <td>338.4995</td>\n",
       "      <td>83.2406</td>\n",
       "      <td>63.1939</td>\n",
       "      <td>50.6896</td>\n",
       "      <td>71.3720</td>\n",
       "      <td>161.8204</td>\n",
       "      <td>204.1681</td>\n",
       "      <td>186.2225</td>\n",
       "      <td>127.6457</td>\n",
       "      <td>172.0572</td>\n",
       "      <td>441.0758</td>\n",
       "      <td>1258.3932</td>\n",
       "      <td>466.5542</td>\n",
       "      <td>450.2848</td>\n",
       "      <td>289.2767</td>\n",
       "      <td>467.7875</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene gene_symbol  GSM2667747  GSM2667748  GSM2667749  \\\n",
       "0  ENSG00000000003      TSPAN6   1766.0677    420.8206    300.8009   \n",
       "1  ENSG00000000005        TNMD     43.9280     18.0091      0.0000   \n",
       "2  ENSG00000000419        DPM1   1097.7855    367.2333    316.5226   \n",
       "3  ENSG00000000457       SCYL3    601.4702    270.9239    163.6869   \n",
       "4  ENSG00000000460    C1orf112   1040.8861    349.1418    298.1800   \n",
       "\n",
       "   GSM2667750  GSM2667751  GSM2667752  GSM2667753  GSM2667754  GSM2667755  \\\n",
       "0   3142.8439   2207.8392   4367.5056    228.5759    650.3287   4880.0185   \n",
       "1    115.8917     63.5936     39.7803     11.7588      1.0153     67.1228   \n",
       "2   3895.0588   1536.1375   1084.5276    127.7205    245.1014   2608.6175   \n",
       "3   1177.2241    873.9074    845.1423    229.8548    134.4694    676.4072   \n",
       "4   1967.3975   2115.2425    980.9359    253.4457    145.7038   1024.4449   \n",
       "\n",
       "   GSM2667756  GSM2667757  GSM2667758  GSM2667759  GSM2667760  GSM2667761  \\\n",
       "0   8629.5404   2951.1930    370.2859    321.4122   3721.4916   3499.9075   \n",
       "1     61.4995     54.1517     11.8111      5.7130     79.6294     79.6162   \n",
       "2   3797.5758    855.0836    124.4127    220.1181   2932.3449   1795.7509   \n",
       "3   1177.2415   1155.1429    232.3545    115.9401   1458.9760   1182.6409   \n",
       "4   1294.0961    599.8307    117.7714     67.3678    785.8567    466.6687   \n",
       "\n",
       "   GSM2667762  GSM2667763  GSM2667764  GSM2667765  GSM2667766  GSM2667767  \\\n",
       "0   2462.5672    300.7646    370.7972   3470.2847   3503.2877   3362.5934   \n",
       "1     69.2199     13.8240      5.9650    138.7558     84.4619    181.6344   \n",
       "2   1499.5849    264.1154    536.7919   3716.6839   1998.1502   1210.7017   \n",
       "3    907.3663    335.9731    153.7599   1083.9457    711.6130   1259.2643   \n",
       "4   1315.1359    300.5638    246.0739   1469.0028   1906.7889   1471.1647   \n",
       "\n",
       "   GSM2667768  GSM2667769  GSM2667770  GSM2667771  GSM2667772  GSM2667773  \\\n",
       "0    226.8240    131.7482   4486.6134   7225.4738   3185.4922    531.7498   \n",
       "1     46.6972     12.8469    168.0059    403.5948     58.7409     18.8835   \n",
       "2    112.5526    150.2928   3655.9428   2458.6382    834.2019    135.4603   \n",
       "3    275.6282     30.2989   1144.3081   1374.0298   1118.6175    325.3986   \n",
       "4    229.3009     67.8212   1176.1864   1307.8388    678.5934    177.1994   \n",
       "\n",
       "   GSM2667774  GSM2667775  GSM2667776  GSM2667777  GSM2667778  GSM2667779  \\\n",
       "0    796.0883   3175.4423   2064.0788   1267.6230   1845.9171    409.7910   \n",
       "1     12.0940     67.2416     10.9099    133.5243    147.7444     15.2011   \n",
       "2    258.5075   2264.3384   1241.8778    217.2588    623.3691    165.4709   \n",
       "3    202.9192   1189.7186    881.3931    124.4839    948.1346     57.5744   \n",
       "4     92.6689    560.3419    386.5186     92.7886    354.3580     49.4149   \n",
       "\n",
       "   GSM2667780  GSM2667781  GSM2667782  GSM2667783  GSM2667784  GSM2667785  \\\n",
       "0   2540.8740   2906.8171   2829.6631   1920.6238   1165.6860   1874.8260   \n",
       "1    202.6725    343.3962    676.6289    376.7828    141.2399    380.3029   \n",
       "2   1166.0064   1281.7182    952.9915    635.3723    879.0727   1695.9609   \n",
       "3    547.4528    439.9557   1454.0821   1793.4598    232.8164    773.3686   \n",
       "4    211.9013    110.9184    673.9635    749.7606     86.1525    266.8906   \n",
       "\n",
       "   GSM2667786  GSM2668081  GSM2668082  GSM2668083  GSM2668084  GSM2668085  \\\n",
       "0   2539.2382    265.0806    300.3062    331.6881    183.7755    278.5022   \n",
       "1    662.0918      0.0000      0.0000      0.0000      0.0000      0.0000   \n",
       "2   1736.9615    537.7681    461.4597    436.1302    416.2508    568.8483   \n",
       "3    939.7315    157.2023    151.5847    196.0139    142.7542    170.0119   \n",
       "4    338.4995     83.2406     63.1939     50.6896     71.3720    161.8204   \n",
       "\n",
       "   GSM2668086  GSM2668087  GSM2668088  GSM2668089  GSM2671001  GSM2671002  \\\n",
       "0    254.7605    143.2263    174.3319    202.3253   1053.8788   2405.3001   \n",
       "1      0.0000      0.0000      0.0000      0.0000      3.9978     11.0025   \n",
       "2    613.1214    374.0965    302.6622    454.0570   1227.1084   2980.0756   \n",
       "3    196.5995    185.3106    260.9485    183.7584    586.7993   1504.0115   \n",
       "4    204.1681    186.2225    127.6457    172.0572    441.0758   1258.3932   \n",
       "\n",
       "   GSM2671003  GSM2671004  GSM2671005  GSM2671006  \n",
       "0    940.5770    973.7475    760.0100   1472.2712  \n",
       "1      4.9979      8.0054      3.9978     11.0011  \n",
       "2   1045.8061   1098.7163    759.7729   1398.8077  \n",
       "3    619.3764    661.9514    469.5312   1113.1412  \n",
       "4    466.5542    450.2848    289.2767    467.7875  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check that it worked\n",
    "\n",
    "pd.set_option('display.max_columns', None) \n",
    "expr_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process gene expression matrix\n",
    "\n",
    "This involves updating Entrez gene ids, sorting and subsetting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(27990, 57)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>GSM2667747</th>\n",
       "      <th>GSM2667748</th>\n",
       "      <th>GSM2667749</th>\n",
       "      <th>GSM2667750</th>\n",
       "      <th>GSM2667751</th>\n",
       "      <th>GSM2667752</th>\n",
       "      <th>GSM2667753</th>\n",
       "      <th>GSM2667754</th>\n",
       "      <th>GSM2667755</th>\n",
       "      <th>GSM2667756</th>\n",
       "      <th>GSM2667757</th>\n",
       "      <th>GSM2667758</th>\n",
       "      <th>GSM2667759</th>\n",
       "      <th>GSM2667760</th>\n",
       "      <th>GSM2667761</th>\n",
       "      <th>GSM2667762</th>\n",
       "      <th>GSM2667763</th>\n",
       "      <th>GSM2667764</th>\n",
       "      <th>GSM2667765</th>\n",
       "      <th>GSM2667766</th>\n",
       "      <th>GSM2667767</th>\n",
       "      <th>GSM2667768</th>\n",
       "      <th>GSM2667769</th>\n",
       "      <th>GSM2667770</th>\n",
       "      <th>GSM2667771</th>\n",
       "      <th>GSM2667772</th>\n",
       "      <th>GSM2667773</th>\n",
       "      <th>GSM2667774</th>\n",
       "      <th>GSM2667775</th>\n",
       "      <th>GSM2667776</th>\n",
       "      <th>GSM2667777</th>\n",
       "      <th>GSM2667778</th>\n",
       "      <th>GSM2667779</th>\n",
       "      <th>GSM2667780</th>\n",
       "      <th>GSM2667781</th>\n",
       "      <th>GSM2667782</th>\n",
       "      <th>GSM2667783</th>\n",
       "      <th>GSM2667784</th>\n",
       "      <th>GSM2667785</th>\n",
       "      <th>GSM2667786</th>\n",
       "      <th>GSM2668081</th>\n",
       "      <th>GSM2668082</th>\n",
       "      <th>GSM2668083</th>\n",
       "      <th>GSM2668084</th>\n",
       "      <th>GSM2668085</th>\n",
       "      <th>GSM2668086</th>\n",
       "      <th>GSM2668087</th>\n",
       "      <th>GSM2668088</th>\n",
       "      <th>GSM2668089</th>\n",
       "      <th>GSM2671001</th>\n",
       "      <th>GSM2671002</th>\n",
       "      <th>GSM2671003</th>\n",
       "      <th>GSM2671004</th>\n",
       "      <th>GSM2671005</th>\n",
       "      <th>GSM2671006</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1766.0677</td>\n",
       "      <td>420.8206</td>\n",
       "      <td>300.8009</td>\n",
       "      <td>3142.8439</td>\n",
       "      <td>2207.8392</td>\n",
       "      <td>4367.5056</td>\n",
       "      <td>228.5759</td>\n",
       "      <td>650.3287</td>\n",
       "      <td>4880.0185</td>\n",
       "      <td>8629.5404</td>\n",
       "      <td>2951.1930</td>\n",
       "      <td>370.2859</td>\n",
       "      <td>321.4122</td>\n",
       "      <td>3721.4916</td>\n",
       "      <td>3499.9075</td>\n",
       "      <td>2462.5672</td>\n",
       "      <td>300.7646</td>\n",
       "      <td>370.7972</td>\n",
       "      <td>3470.2847</td>\n",
       "      <td>3503.2877</td>\n",
       "      <td>3362.5934</td>\n",
       "      <td>226.8240</td>\n",
       "      <td>131.7482</td>\n",
       "      <td>4486.6134</td>\n",
       "      <td>7225.4738</td>\n",
       "      <td>3185.4922</td>\n",
       "      <td>531.7498</td>\n",
       "      <td>796.0883</td>\n",
       "      <td>3175.4423</td>\n",
       "      <td>2064.0788</td>\n",
       "      <td>1267.6230</td>\n",
       "      <td>1845.9171</td>\n",
       "      <td>409.7910</td>\n",
       "      <td>2540.8740</td>\n",
       "      <td>2906.8171</td>\n",
       "      <td>2829.6631</td>\n",
       "      <td>1920.6238</td>\n",
       "      <td>1165.6860</td>\n",
       "      <td>1874.8260</td>\n",
       "      <td>2539.2382</td>\n",
       "      <td>265.0806</td>\n",
       "      <td>300.3062</td>\n",
       "      <td>331.6881</td>\n",
       "      <td>183.7755</td>\n",
       "      <td>278.5022</td>\n",
       "      <td>254.7605</td>\n",
       "      <td>143.2263</td>\n",
       "      <td>174.3319</td>\n",
       "      <td>202.3253</td>\n",
       "      <td>1053.8788</td>\n",
       "      <td>2405.3001</td>\n",
       "      <td>940.5770</td>\n",
       "      <td>973.7475</td>\n",
       "      <td>760.0100</td>\n",
       "      <td>1472.2712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>43.9280</td>\n",
       "      <td>18.0091</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>115.8917</td>\n",
       "      <td>63.5936</td>\n",
       "      <td>39.7803</td>\n",
       "      <td>11.7588</td>\n",
       "      <td>1.0153</td>\n",
       "      <td>67.1228</td>\n",
       "      <td>61.4995</td>\n",
       "      <td>54.1517</td>\n",
       "      <td>11.8111</td>\n",
       "      <td>5.7130</td>\n",
       "      <td>79.6294</td>\n",
       "      <td>79.6162</td>\n",
       "      <td>69.2199</td>\n",
       "      <td>13.8240</td>\n",
       "      <td>5.9650</td>\n",
       "      <td>138.7558</td>\n",
       "      <td>84.4619</td>\n",
       "      <td>181.6344</td>\n",
       "      <td>46.6972</td>\n",
       "      <td>12.8469</td>\n",
       "      <td>168.0059</td>\n",
       "      <td>403.5948</td>\n",
       "      <td>58.7409</td>\n",
       "      <td>18.8835</td>\n",
       "      <td>12.0940</td>\n",
       "      <td>67.2416</td>\n",
       "      <td>10.9099</td>\n",
       "      <td>133.5243</td>\n",
       "      <td>147.7444</td>\n",
       "      <td>15.2011</td>\n",
       "      <td>202.6725</td>\n",
       "      <td>343.3962</td>\n",
       "      <td>676.6289</td>\n",
       "      <td>376.7828</td>\n",
       "      <td>141.2399</td>\n",
       "      <td>380.3029</td>\n",
       "      <td>662.0918</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3.9978</td>\n",
       "      <td>11.0025</td>\n",
       "      <td>4.9979</td>\n",
       "      <td>8.0054</td>\n",
       "      <td>3.9978</td>\n",
       "      <td>11.0011</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      gene gene_symbol  GSM2667747  GSM2667748  GSM2667749  \\\n",
       "sample_id                                                                    \n",
       "0          ENSG00000000003      TSPAN6   1766.0677    420.8206    300.8009   \n",
       "1          ENSG00000000005        TNMD     43.9280     18.0091      0.0000   \n",
       "\n",
       "           GSM2667750  GSM2667751  GSM2667752  GSM2667753  GSM2667754  \\\n",
       "sample_id                                                               \n",
       "0           3142.8439   2207.8392   4367.5056    228.5759    650.3287   \n",
       "1            115.8917     63.5936     39.7803     11.7588      1.0153   \n",
       "\n",
       "           GSM2667755  GSM2667756  GSM2667757  GSM2667758  GSM2667759  \\\n",
       "sample_id                                                               \n",
       "0           4880.0185   8629.5404   2951.1930    370.2859    321.4122   \n",
       "1             67.1228     61.4995     54.1517     11.8111      5.7130   \n",
       "\n",
       "           GSM2667760  GSM2667761  GSM2667762  GSM2667763  GSM2667764  \\\n",
       "sample_id                                                               \n",
       "0           3721.4916   3499.9075   2462.5672    300.7646    370.7972   \n",
       "1             79.6294     79.6162     69.2199     13.8240      5.9650   \n",
       "\n",
       "           GSM2667765  GSM2667766  GSM2667767  GSM2667768  GSM2667769  \\\n",
       "sample_id                                                               \n",
       "0           3470.2847   3503.2877   3362.5934    226.8240    131.7482   \n",
       "1            138.7558     84.4619    181.6344     46.6972     12.8469   \n",
       "\n",
       "           GSM2667770  GSM2667771  GSM2667772  GSM2667773  GSM2667774  \\\n",
       "sample_id                                                               \n",
       "0           4486.6134   7225.4738   3185.4922    531.7498    796.0883   \n",
       "1            168.0059    403.5948     58.7409     18.8835     12.0940   \n",
       "\n",
       "           GSM2667775  GSM2667776  GSM2667777  GSM2667778  GSM2667779  \\\n",
       "sample_id                                                               \n",
       "0           3175.4423   2064.0788   1267.6230   1845.9171    409.7910   \n",
       "1             67.2416     10.9099    133.5243    147.7444     15.2011   \n",
       "\n",
       "           GSM2667780  GSM2667781  GSM2667782  GSM2667783  GSM2667784  \\\n",
       "sample_id                                                               \n",
       "0           2540.8740   2906.8171   2829.6631   1920.6238   1165.6860   \n",
       "1            202.6725    343.3962    676.6289    376.7828    141.2399   \n",
       "\n",
       "           GSM2667785  GSM2667786  GSM2668081  GSM2668082  GSM2668083  \\\n",
       "sample_id                                                               \n",
       "0           1874.8260   2539.2382    265.0806    300.3062    331.6881   \n",
       "1            380.3029    662.0918      0.0000      0.0000      0.0000   \n",
       "\n",
       "           GSM2668084  GSM2668085  GSM2668086  GSM2668087  GSM2668088  \\\n",
       "sample_id                                                               \n",
       "0            183.7755    278.5022    254.7605    143.2263    174.3319   \n",
       "1              0.0000      0.0000      0.0000      0.0000      0.0000   \n",
       "\n",
       "           GSM2668089  GSM2671001  GSM2671002  GSM2671003  GSM2671004  \\\n",
       "sample_id                                                               \n",
       "0            202.3253   1053.8788   2405.3001    940.5770    973.7475   \n",
       "1              0.0000      3.9978     11.0025      4.9979      8.0054   \n",
       "\n",
       "           GSM2671005  GSM2671006  \n",
       "sample_id                          \n",
       "0            760.0100   1472.2712  \n",
       "1              3.9978     11.0011  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# expr_df = (expr_df\n",
    "#     .dropna(axis='rows')\n",
    "#     .reindex(probe_map_df.id)\n",
    "#     .rename(index=ensembl_to_entrez)\n",
    "#     .rename(index=old_to_new_entrez)\n",
    "#     .groupby(level=0).mean()\n",
    "#     .transpose()\n",
    "#     .sort_index(axis='rows')\n",
    "#     .sort_index(axis='columns')\n",
    "# )\n",
    "\n",
    "expr_df.index.rename('sample_id', inplace=True)\n",
    "\n",
    "print(expr_df.shape)\n",
    "expr_df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stratify Balanced Training and Testing Sets in TARGET Gene Expression\n",
    "\n",
    "Output training and testing gene expression datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#strat = pheno_df.set_index('sample_id').reindex(expr_df.index).primary_disease_code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cancertype_count_df = (\n",
    "#     pd.DataFrame(strat.value_counts()) #not using value_counts, what number should we use here? \n",
    "#     .reset_index()\n",
    "#     .rename({'index': 'cancertype', 'primary_disease_code': 'n ='}, axis='columns')\n",
    "# )\n",
    "\n",
    "# file = os.path.join('data', 'target_sample_counts.tsv') #change which file - do we have a file that works for this?\n",
    "# cancertype_count_df.to_csv(file, sep='\\t', index=False)\n",
    "\n",
    "# cancertype_count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df, test_df = train_test_split(expr_df,\n",
    "                                     test_size=0.1,\n",
    "                                     random_state=123) #if no stratify defined, should just randomize on its own"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(25191, 57)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(2799, 57)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(train_df.shape)\n",
    "test_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save train dataframe to file \n",
    "train_file = os.path.join('data', 'train_grein_human_expression_matrix_processed.tsv.gz')\n",
    "train_df.to_csv(train_file, sep='\\t', compression='gzip', float_format='%.3g')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save test dataframe to file \n",
    "test_file = os.path.join('data', 'test_grein_human_expression_matrix_processed.tsv.gz')\n",
    "test_df.to_csv(test_file, sep='\\t', compression='gzip', float_format='%.3g')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sort genes based on median absolute deviation and output to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to calculate median absolute deviation\n",
    "def mad(df): \n",
    "    \"\"\"Function to calculate median absolute deviation for a dataframe\n",
    "  argument1 (dataframe): Dataframe for which to calculate the median absolute deviation, row by row\n",
    "\n",
    "  Returns: The median absolute deviation for each row of the data frame\n",
    "  \"\"\"\n",
    "    row_medians = df.median(axis='columns', numeric_only=True)\n",
    "    abs_row_median_diffs = abs(df.sub(row_medians, axis='rows'))\n",
    "    return abs_row_median_diffs.median(axis='columns', numeric_only=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>GSM2667747</th>\n",
       "      <th>GSM2667748</th>\n",
       "      <th>GSM2667749</th>\n",
       "      <th>GSM2667750</th>\n",
       "      <th>GSM2667751</th>\n",
       "      <th>GSM2667752</th>\n",
       "      <th>GSM2667753</th>\n",
       "      <th>GSM2667754</th>\n",
       "      <th>GSM2667755</th>\n",
       "      <th>GSM2667756</th>\n",
       "      <th>GSM2667757</th>\n",
       "      <th>GSM2667758</th>\n",
       "      <th>GSM2667759</th>\n",
       "      <th>GSM2667760</th>\n",
       "      <th>GSM2667761</th>\n",
       "      <th>GSM2667762</th>\n",
       "      <th>GSM2667763</th>\n",
       "      <th>GSM2667764</th>\n",
       "      <th>GSM2667765</th>\n",
       "      <th>GSM2667766</th>\n",
       "      <th>GSM2667767</th>\n",
       "      <th>GSM2667768</th>\n",
       "      <th>GSM2667769</th>\n",
       "      <th>GSM2667770</th>\n",
       "      <th>GSM2667771</th>\n",
       "      <th>GSM2667772</th>\n",
       "      <th>GSM2667773</th>\n",
       "      <th>GSM2667774</th>\n",
       "      <th>GSM2667775</th>\n",
       "      <th>GSM2667776</th>\n",
       "      <th>GSM2667777</th>\n",
       "      <th>GSM2667778</th>\n",
       "      <th>GSM2667779</th>\n",
       "      <th>GSM2667780</th>\n",
       "      <th>GSM2667781</th>\n",
       "      <th>GSM2667782</th>\n",
       "      <th>GSM2667783</th>\n",
       "      <th>GSM2667784</th>\n",
       "      <th>GSM2667785</th>\n",
       "      <th>GSM2667786</th>\n",
       "      <th>GSM2668081</th>\n",
       "      <th>GSM2668082</th>\n",
       "      <th>GSM2668083</th>\n",
       "      <th>GSM2668084</th>\n",
       "      <th>GSM2668085</th>\n",
       "      <th>GSM2668086</th>\n",
       "      <th>GSM2668087</th>\n",
       "      <th>GSM2668088</th>\n",
       "      <th>GSM2668089</th>\n",
       "      <th>GSM2671001</th>\n",
       "      <th>GSM2671002</th>\n",
       "      <th>GSM2671003</th>\n",
       "      <th>GSM2671004</th>\n",
       "      <th>GSM2671005</th>\n",
       "      <th>GSM2671006</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27744</th>\n",
       "      <td>ENSG00000283046HTN1</td>\n",
       "      <td>ENSG00000283046</td>\n",
       "      <td>HTN1</td>\n",
       "      <td>125.4665</td>\n",
       "      <td>140.2465</td>\n",
       "      <td>17.6274</td>\n",
       "      <td>69.7745</td>\n",
       "      <td>73.7733</td>\n",
       "      <td>27.6203</td>\n",
       "      <td>98.1395</td>\n",
       "      <td>12.7103</td>\n",
       "      <td>3.6260</td>\n",
       "      <td>24.9714</td>\n",
       "      <td>51.9728</td>\n",
       "      <td>66.2000</td>\n",
       "      <td>3.0588</td>\n",
       "      <td>29.6436</td>\n",
       "      <td>78.6089</td>\n",
       "      <td>192.3800</td>\n",
       "      <td>153.2949</td>\n",
       "      <td>4.0605</td>\n",
       "      <td>61.9086</td>\n",
       "      <td>132.9720</td>\n",
       "      <td>178.1983</td>\n",
       "      <td>117.1625</td>\n",
       "      <td>6.5244</td>\n",
       "      <td>18.5502</td>\n",
       "      <td>57.2569</td>\n",
       "      <td>106.8320</td>\n",
       "      <td>148.3160</td>\n",
       "      <td>14.3281</td>\n",
       "      <td>19.2520</td>\n",
       "      <td>147.4773</td>\n",
       "      <td>28.1115</td>\n",
       "      <td>281.7885</td>\n",
       "      <td>1.5309</td>\n",
       "      <td>18.0041</td>\n",
       "      <td>14.8531</td>\n",
       "      <td>305.5563</td>\n",
       "      <td>848.8558</td>\n",
       "      <td>13.7288</td>\n",
       "      <td>23.7737</td>\n",
       "      <td>30.6058</td>\n",
       "      <td>5.7421</td>\n",
       "      <td>4.7490</td>\n",
       "      <td>7.7266</td>\n",
       "      <td>1.0890</td>\n",
       "      <td>3.8178</td>\n",
       "      <td>8.0805</td>\n",
       "      <td>5.0509</td>\n",
       "      <td>7.7629</td>\n",
       "      <td>7.2611</td>\n",
       "      <td>58.6780</td>\n",
       "      <td>185.7149</td>\n",
       "      <td>86.5948</td>\n",
       "      <td>123.6247</td>\n",
       "      <td>97.2255</td>\n",
       "      <td>128.9266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9537</th>\n",
       "      <td>ENSG00000153037SRP19</td>\n",
       "      <td>ENSG00000153037</td>\n",
       "      <td>SRP19</td>\n",
       "      <td>894.5558</td>\n",
       "      <td>357.1081</td>\n",
       "      <td>187.4946</td>\n",
       "      <td>1421.0479</td>\n",
       "      <td>456.0143</td>\n",
       "      <td>802.6670</td>\n",
       "      <td>124.8560</td>\n",
       "      <td>164.7368</td>\n",
       "      <td>3532.8830</td>\n",
       "      <td>1244.8740</td>\n",
       "      <td>1010.8748</td>\n",
       "      <td>139.9850</td>\n",
       "      <td>138.3330</td>\n",
       "      <td>3899.5298</td>\n",
       "      <td>930.8268</td>\n",
       "      <td>1137.9553</td>\n",
       "      <td>254.7177</td>\n",
       "      <td>245.6310</td>\n",
       "      <td>1621.9018</td>\n",
       "      <td>456.3314</td>\n",
       "      <td>1035.1774</td>\n",
       "      <td>116.1225</td>\n",
       "      <td>123.1838</td>\n",
       "      <td>2368.7324</td>\n",
       "      <td>921.9137</td>\n",
       "      <td>1162.6473</td>\n",
       "      <td>218.2426</td>\n",
       "      <td>274.7572</td>\n",
       "      <td>2619.1004</td>\n",
       "      <td>674.3049</td>\n",
       "      <td>181.9868</td>\n",
       "      <td>715.1966</td>\n",
       "      <td>68.2822</td>\n",
       "      <td>985.3323</td>\n",
       "      <td>772.8154</td>\n",
       "      <td>1533.7694</td>\n",
       "      <td>953.2965</td>\n",
       "      <td>784.0392</td>\n",
       "      <td>2934.4142</td>\n",
       "      <td>1425.9365</td>\n",
       "      <td>598.9880</td>\n",
       "      <td>577.2314</td>\n",
       "      <td>684.2199</td>\n",
       "      <td>376.1889</td>\n",
       "      <td>521.7010</td>\n",
       "      <td>656.0729</td>\n",
       "      <td>793.1388</td>\n",
       "      <td>850.7232</td>\n",
       "      <td>926.3904</td>\n",
       "      <td>642.6850</td>\n",
       "      <td>1674.6958</td>\n",
       "      <td>614.6459</td>\n",
       "      <td>749.6716</td>\n",
       "      <td>533.6487</td>\n",
       "      <td>943.2409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9769</th>\n",
       "      <td>ENSG00000155115GTF3C6</td>\n",
       "      <td>ENSG00000155115</td>\n",
       "      <td>GTF3C6</td>\n",
       "      <td>851.6143</td>\n",
       "      <td>187.0203</td>\n",
       "      <td>243.4404</td>\n",
       "      <td>2046.5828</td>\n",
       "      <td>476.0532</td>\n",
       "      <td>1977.8905</td>\n",
       "      <td>94.7822</td>\n",
       "      <td>881.8821</td>\n",
       "      <td>10161.3111</td>\n",
       "      <td>1382.6290</td>\n",
       "      <td>1326.2597</td>\n",
       "      <td>141.7311</td>\n",
       "      <td>477.2282</td>\n",
       "      <td>5214.7134</td>\n",
       "      <td>767.6003</td>\n",
       "      <td>982.4279</td>\n",
       "      <td>117.6392</td>\n",
       "      <td>369.8380</td>\n",
       "      <td>2667.1936</td>\n",
       "      <td>459.5399</td>\n",
       "      <td>1652.4284</td>\n",
       "      <td>63.1625</td>\n",
       "      <td>275.7449</td>\n",
       "      <td>3911.4982</td>\n",
       "      <td>1060.2774</td>\n",
       "      <td>1230.0665</td>\n",
       "      <td>147.4217</td>\n",
       "      <td>466.3677</td>\n",
       "      <td>4542.0178</td>\n",
       "      <td>614.1263</td>\n",
       "      <td>202.8739</td>\n",
       "      <td>289.3384</td>\n",
       "      <td>166.5569</td>\n",
       "      <td>1629.0443</td>\n",
       "      <td>566.4231</td>\n",
       "      <td>1570.2533</td>\n",
       "      <td>625.5507</td>\n",
       "      <td>2881.6804</td>\n",
       "      <td>3669.3315</td>\n",
       "      <td>1164.9276</td>\n",
       "      <td>777.0025</td>\n",
       "      <td>705.8804</td>\n",
       "      <td>790.7663</td>\n",
       "      <td>536.9289</td>\n",
       "      <td>669.2589</td>\n",
       "      <td>755.7290</td>\n",
       "      <td>827.7310</td>\n",
       "      <td>785.4842</td>\n",
       "      <td>994.0243</td>\n",
       "      <td>268.1323</td>\n",
       "      <td>567.9704</td>\n",
       "      <td>218.9266</td>\n",
       "      <td>288.9725</td>\n",
       "      <td>250.9841</td>\n",
       "      <td>536.0405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15239</th>\n",
       "      <td>ENSG00000184083FAM120C</td>\n",
       "      <td>ENSG00000184083</td>\n",
       "      <td>FAM120C</td>\n",
       "      <td>632.3881</td>\n",
       "      <td>285.1058</td>\n",
       "      <td>122.8606</td>\n",
       "      <td>832.8702</td>\n",
       "      <td>957.4550</td>\n",
       "      <td>1495.5920</td>\n",
       "      <td>184.7934</td>\n",
       "      <td>91.7322</td>\n",
       "      <td>383.1359</td>\n",
       "      <td>1374.5838</td>\n",
       "      <td>1628.9039</td>\n",
       "      <td>232.1640</td>\n",
       "      <td>120.3054</td>\n",
       "      <td>898.8325</td>\n",
       "      <td>1279.4379</td>\n",
       "      <td>824.3853</td>\n",
       "      <td>188.6935</td>\n",
       "      <td>166.3629</td>\n",
       "      <td>881.8608</td>\n",
       "      <td>1010.1709</td>\n",
       "      <td>1574.6568</td>\n",
       "      <td>261.7398</td>\n",
       "      <td>58.3872</td>\n",
       "      <td>756.5876</td>\n",
       "      <td>2166.1719</td>\n",
       "      <td>1840.3923</td>\n",
       "      <td>257.8297</td>\n",
       "      <td>213.4336</td>\n",
       "      <td>876.4116</td>\n",
       "      <td>1554.1902</td>\n",
       "      <td>507.1274</td>\n",
       "      <td>1001.3734</td>\n",
       "      <td>76.3392</td>\n",
       "      <td>493.3627</td>\n",
       "      <td>725.1371</td>\n",
       "      <td>2403.1399</td>\n",
       "      <td>1715.6976</td>\n",
       "      <td>344.4401</td>\n",
       "      <td>526.2655</td>\n",
       "      <td>1218.9705</td>\n",
       "      <td>80.7898</td>\n",
       "      <td>72.3151</td>\n",
       "      <td>55.0190</td>\n",
       "      <td>147.1572</td>\n",
       "      <td>131.1924</td>\n",
       "      <td>162.8560</td>\n",
       "      <td>110.0482</td>\n",
       "      <td>141.0302</td>\n",
       "      <td>109.8994</td>\n",
       "      <td>159.4169</td>\n",
       "      <td>277.8073</td>\n",
       "      <td>134.6019</td>\n",
       "      <td>252.2688</td>\n",
       "      <td>188.7017</td>\n",
       "      <td>352.3888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>ENSG00000008196TFAP2B</td>\n",
       "      <td>ENSG00000008196</td>\n",
       "      <td>TFAP2B</td>\n",
       "      <td>23.2873</td>\n",
       "      <td>2.0018</td>\n",
       "      <td>19.1983</td>\n",
       "      <td>54.7994</td>\n",
       "      <td>19.0940</td>\n",
       "      <td>974.5680</td>\n",
       "      <td>88.7088</td>\n",
       "      <td>88.8378</td>\n",
       "      <td>620.9386</td>\n",
       "      <td>1124.1208</td>\n",
       "      <td>6430.8975</td>\n",
       "      <td>744.3863</td>\n",
       "      <td>499.9203</td>\n",
       "      <td>5383.4356</td>\n",
       "      <td>7004.7640</td>\n",
       "      <td>18.7766</td>\n",
       "      <td>6.8329</td>\n",
       "      <td>7.8788</td>\n",
       "      <td>36.7517</td>\n",
       "      <td>19.2128</td>\n",
       "      <td>1565.6738</td>\n",
       "      <td>132.5760</td>\n",
       "      <td>22.7705</td>\n",
       "      <td>1405.0990</td>\n",
       "      <td>2319.8399</td>\n",
       "      <td>7436.2614</td>\n",
       "      <td>813.1225</td>\n",
       "      <td>864.8631</td>\n",
       "      <td>4407.3383</td>\n",
       "      <td>5406.7102</td>\n",
       "      <td>439.1032</td>\n",
       "      <td>1307.2588</td>\n",
       "      <td>149.5145</td>\n",
       "      <td>998.4512</td>\n",
       "      <td>1370.7760</td>\n",
       "      <td>4782.1691</td>\n",
       "      <td>2161.5760</td>\n",
       "      <td>1078.2573</td>\n",
       "      <td>2003.7820</td>\n",
       "      <td>3485.0708</td>\n",
       "      <td>51.0392</td>\n",
       "      <td>55.8615</td>\n",
       "      <td>39.7806</td>\n",
       "      <td>27.1175</td>\n",
       "      <td>30.9595</td>\n",
       "      <td>31.9595</td>\n",
       "      <td>120.6547</td>\n",
       "      <td>156.4436</td>\n",
       "      <td>155.0772</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          gene_id             gene gene_symbol  GSM2667747  \\\n",
       "sample_id                                                                    \n",
       "27744         ENSG00000283046HTN1  ENSG00000283046        HTN1    125.4665   \n",
       "9537         ENSG00000153037SRP19  ENSG00000153037       SRP19    894.5558   \n",
       "9769        ENSG00000155115GTF3C6  ENSG00000155115      GTF3C6    851.6143   \n",
       "15239      ENSG00000184083FAM120C  ENSG00000184083     FAM120C    632.3881   \n",
       "203         ENSG00000008196TFAP2B  ENSG00000008196      TFAP2B     23.2873   \n",
       "\n",
       "           GSM2667748  GSM2667749  GSM2667750  GSM2667751  GSM2667752  \\\n",
       "sample_id                                                               \n",
       "27744        140.2465     17.6274     69.7745     73.7733     27.6203   \n",
       "9537         357.1081    187.4946   1421.0479    456.0143    802.6670   \n",
       "9769         187.0203    243.4404   2046.5828    476.0532   1977.8905   \n",
       "15239        285.1058    122.8606    832.8702    957.4550   1495.5920   \n",
       "203            2.0018     19.1983     54.7994     19.0940    974.5680   \n",
       "\n",
       "           GSM2667753  GSM2667754  GSM2667755  GSM2667756  GSM2667757  \\\n",
       "sample_id                                                               \n",
       "27744         98.1395     12.7103      3.6260     24.9714     51.9728   \n",
       "9537         124.8560    164.7368   3532.8830   1244.8740   1010.8748   \n",
       "9769          94.7822    881.8821  10161.3111   1382.6290   1326.2597   \n",
       "15239        184.7934     91.7322    383.1359   1374.5838   1628.9039   \n",
       "203           88.7088     88.8378    620.9386   1124.1208   6430.8975   \n",
       "\n",
       "           GSM2667758  GSM2667759  GSM2667760  GSM2667761  GSM2667762  \\\n",
       "sample_id                                                               \n",
       "27744         66.2000      3.0588     29.6436     78.6089    192.3800   \n",
       "9537         139.9850    138.3330   3899.5298    930.8268   1137.9553   \n",
       "9769         141.7311    477.2282   5214.7134    767.6003    982.4279   \n",
       "15239        232.1640    120.3054    898.8325   1279.4379    824.3853   \n",
       "203          744.3863    499.9203   5383.4356   7004.7640     18.7766   \n",
       "\n",
       "           GSM2667763  GSM2667764  GSM2667765  GSM2667766  GSM2667767  \\\n",
       "sample_id                                                               \n",
       "27744        153.2949      4.0605     61.9086    132.9720    178.1983   \n",
       "9537         254.7177    245.6310   1621.9018    456.3314   1035.1774   \n",
       "9769         117.6392    369.8380   2667.1936    459.5399   1652.4284   \n",
       "15239        188.6935    166.3629    881.8608   1010.1709   1574.6568   \n",
       "203            6.8329      7.8788     36.7517     19.2128   1565.6738   \n",
       "\n",
       "           GSM2667768  GSM2667769  GSM2667770  GSM2667771  GSM2667772  \\\n",
       "sample_id                                                               \n",
       "27744        117.1625      6.5244     18.5502     57.2569    106.8320   \n",
       "9537         116.1225    123.1838   2368.7324    921.9137   1162.6473   \n",
       "9769          63.1625    275.7449   3911.4982   1060.2774   1230.0665   \n",
       "15239        261.7398     58.3872    756.5876   2166.1719   1840.3923   \n",
       "203          132.5760     22.7705   1405.0990   2319.8399   7436.2614   \n",
       "\n",
       "           GSM2667773  GSM2667774  GSM2667775  GSM2667776  GSM2667777  \\\n",
       "sample_id                                                               \n",
       "27744        148.3160     14.3281     19.2520    147.4773     28.1115   \n",
       "9537         218.2426    274.7572   2619.1004    674.3049    181.9868   \n",
       "9769         147.4217    466.3677   4542.0178    614.1263    202.8739   \n",
       "15239        257.8297    213.4336    876.4116   1554.1902    507.1274   \n",
       "203          813.1225    864.8631   4407.3383   5406.7102    439.1032   \n",
       "\n",
       "           GSM2667778  GSM2667779  GSM2667780  GSM2667781  GSM2667782  \\\n",
       "sample_id                                                               \n",
       "27744        281.7885      1.5309     18.0041     14.8531    305.5563   \n",
       "9537         715.1966     68.2822    985.3323    772.8154   1533.7694   \n",
       "9769         289.3384    166.5569   1629.0443    566.4231   1570.2533   \n",
       "15239       1001.3734     76.3392    493.3627    725.1371   2403.1399   \n",
       "203         1307.2588    149.5145    998.4512   1370.7760   4782.1691   \n",
       "\n",
       "           GSM2667783  GSM2667784  GSM2667785  GSM2667786  GSM2668081  \\\n",
       "sample_id                                                               \n",
       "27744        848.8558     13.7288     23.7737     30.6058      5.7421   \n",
       "9537         953.2965    784.0392   2934.4142   1425.9365    598.9880   \n",
       "9769         625.5507   2881.6804   3669.3315   1164.9276    777.0025   \n",
       "15239       1715.6976    344.4401    526.2655   1218.9705     80.7898   \n",
       "203         2161.5760   1078.2573   2003.7820   3485.0708     51.0392   \n",
       "\n",
       "           GSM2668082  GSM2668083  GSM2668084  GSM2668085  GSM2668086  \\\n",
       "sample_id                                                               \n",
       "27744          4.7490      7.7266      1.0890      3.8178      8.0805   \n",
       "9537         577.2314    684.2199    376.1889    521.7010    656.0729   \n",
       "9769         705.8804    790.7663    536.9289    669.2589    755.7290   \n",
       "15239         72.3151     55.0190    147.1572    131.1924    162.8560   \n",
       "203           55.8615     39.7806     27.1175     30.9595     31.9595   \n",
       "\n",
       "           GSM2668087  GSM2668088  GSM2668089  GSM2671001  GSM2671002  \\\n",
       "sample_id                                                               \n",
       "27744          5.0509      7.7629      7.2611     58.6780    185.7149   \n",
       "9537         793.1388    850.7232    926.3904    642.6850   1674.6958   \n",
       "9769         827.7310    785.4842    994.0243    268.1323    567.9704   \n",
       "15239        110.0482    141.0302    109.8994    159.4169    277.8073   \n",
       "203          120.6547    156.4436    155.0772      0.0000      0.0000   \n",
       "\n",
       "           GSM2671003  GSM2671004  GSM2671005  GSM2671006  \n",
       "sample_id                                                  \n",
       "27744         86.5948    123.6247     97.2255    128.9266  \n",
       "9537         614.6459    749.6716    533.6487    943.2409  \n",
       "9769         218.9266    288.9725    250.9841    536.0405  \n",
       "15239        134.6019    252.2688    188.7017    352.3888  \n",
       "203            0.0000      0.0000      0.0000      0.0000  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add gene_id as a column in dataframe\n",
    "train_df['gene_id'] = train_df['gene'] + train_df['gene_symbol']\n",
    "col = train_df.pop('gene_id')\n",
    "train_df.insert(0, col.name, col)\n",
    "\n",
    "train_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample_id\n",
       "26434    2.318332e+06\n",
       "25781    2.305465e+06\n",
       "27829    4.295272e+05\n",
       "26724    1.809434e+05\n",
       "9917     1.733389e+05\n",
       "             ...     \n",
       "8967     0.000000e+00\n",
       "26441    0.000000e+00\n",
       "26418    0.000000e+00\n",
       "27924    0.000000e+00\n",
       "19966    0.000000e+00\n",
       "Length: 25191, dtype: float64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Determine most variably expressed genes and subset\n",
    "# create dataframe to save median absolute deviation data for rats\n",
    "train_df_mad = mad(train_df.drop(['gene_id','gene', 'gene_symbol'], axis = 1))\n",
    "train_df_mad.columns = ['gene_id', 'median_abs_deviation']\n",
    "\n",
    "train_df_mad = train_df_mad.sort_values(ascending=False)\n",
    "train_df_mad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to tsv file\n",
    "file = os.path.join('data', 'grein_mad_human_genes.tsv')\n",
    "train_df_mad.to_csv(file, sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
