{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process TARGET PanCancer Data\n",
    "\n",
    "Retrieve the downloaded expression data, update gene identifiers to entrez, and curate sample IDs. The script will also identify a balanced hold-out test set to compare projection performance into learned latent spaces across algorithms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import random\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(1234)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Phenotype Information: skipping until phenotype data found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path = os.path.join('download', 'TARGET_phenotype.gz')\n",
    "# pheno_df = pd.read_table(path)\n",
    "\n",
    "# print(pheno_df.shape)\n",
    "# pheno_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Probe Mapping Info \n",
    "(where chromosomes start and end, saved in a file in downloads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path = os.path.join('download', 'gencode.v23.annotation.gene.probemap')\n",
    "# probe_map_df = pd.read_table(path)\n",
    "\n",
    "# # Inner merge gene df to get ensembl to entrez mapping\n",
    "# probe_map_df = probe_map_df.merge(gene_df, how='inner', left_on='gene', right_on='symbol')\n",
    "\n",
    "# # Mapping to rename gene expression index\n",
    "# ensembl_to_entrez = dict(zip(probe_map_df.id, probe_map_df.entrez_gene_id))\n",
    "\n",
    "# print(probe_map_df.shape)\n",
    "# probe_map_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Gene Expression Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(27990, 1635)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>GSM2667747_x</th>\n",
       "      <th>GSM2667748_x</th>\n",
       "      <th>GSM2667749_x</th>\n",
       "      <th>GSM2667750_x</th>\n",
       "      <th>GSM2667751_x</th>\n",
       "      <th>GSM2667752_x</th>\n",
       "      <th>GSM2667753_x</th>\n",
       "      <th>GSM2667754_x</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2701937</th>\n",
       "      <th>GSM2701938</th>\n",
       "      <th>GSM2701939</th>\n",
       "      <th>GSM2701940</th>\n",
       "      <th>GSM2701941</th>\n",
       "      <th>GSM2701942</th>\n",
       "      <th>GSM2701943</th>\n",
       "      <th>GSM2701944</th>\n",
       "      <th>GSM2701945</th>\n",
       "      <th>GSM2701946</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1766.0677</td>\n",
       "      <td>420.8206</td>\n",
       "      <td>300.8009</td>\n",
       "      <td>3142.8439</td>\n",
       "      <td>2207.8392</td>\n",
       "      <td>4367.5056</td>\n",
       "      <td>228.5759</td>\n",
       "      <td>650.3287</td>\n",
       "      <td>...</td>\n",
       "      <td>17.4324</td>\n",
       "      <td>13.7006</td>\n",
       "      <td>9.8935</td>\n",
       "      <td>11.9023</td>\n",
       "      <td>11.3564</td>\n",
       "      <td>8.9213</td>\n",
       "      <td>16.4894</td>\n",
       "      <td>13.9479</td>\n",
       "      <td>16.8066</td>\n",
       "      <td>7.7069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>43.9280</td>\n",
       "      <td>18.0091</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>115.8917</td>\n",
       "      <td>63.5936</td>\n",
       "      <td>39.7803</td>\n",
       "      <td>11.7588</td>\n",
       "      <td>1.0153</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0000</td>\n",
       "      <td>2.0443</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>5.5140</td>\n",
       "      <td>7.6165</td>\n",
       "      <td>6.6037</td>\n",
       "      <td>3.0000</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>4.5765</td>\n",
       "      <td>1.0551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>1097.7855</td>\n",
       "      <td>367.2333</td>\n",
       "      <td>316.5226</td>\n",
       "      <td>3895.0588</td>\n",
       "      <td>1536.1375</td>\n",
       "      <td>1084.5276</td>\n",
       "      <td>127.7205</td>\n",
       "      <td>245.1014</td>\n",
       "      <td>...</td>\n",
       "      <td>165.9822</td>\n",
       "      <td>217.9993</td>\n",
       "      <td>114.9995</td>\n",
       "      <td>307.4970</td>\n",
       "      <td>272.9994</td>\n",
       "      <td>301.4232</td>\n",
       "      <td>211.9566</td>\n",
       "      <td>227.9990</td>\n",
       "      <td>264.4285</td>\n",
       "      <td>256.9996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>601.4702</td>\n",
       "      <td>270.9239</td>\n",
       "      <td>163.6869</td>\n",
       "      <td>1177.2241</td>\n",
       "      <td>873.9074</td>\n",
       "      <td>845.1423</td>\n",
       "      <td>229.8548</td>\n",
       "      <td>134.4694</td>\n",
       "      <td>...</td>\n",
       "      <td>96.6508</td>\n",
       "      <td>84.5137</td>\n",
       "      <td>70.8997</td>\n",
       "      <td>99.7864</td>\n",
       "      <td>145.2329</td>\n",
       "      <td>122.9726</td>\n",
       "      <td>81.5081</td>\n",
       "      <td>79.0215</td>\n",
       "      <td>99.9146</td>\n",
       "      <td>77.1152</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows Ã— 1635 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene gene_symbol  GSM2667747_x  GSM2667748_x  GSM2667749_x  \\\n",
       "0  ENSG00000000003      TSPAN6     1766.0677      420.8206      300.8009   \n",
       "1  ENSG00000000005        TNMD       43.9280       18.0091        0.0000   \n",
       "2  ENSG00000000419        DPM1     1097.7855      367.2333      316.5226   \n",
       "3  ENSG00000000457       SCYL3      601.4702      270.9239      163.6869   \n",
       "\n",
       "   GSM2667750_x  GSM2667751_x  GSM2667752_x  GSM2667753_x  GSM2667754_x  ...  \\\n",
       "0     3142.8439     2207.8392     4367.5056      228.5759      650.3287  ...   \n",
       "1      115.8917       63.5936       39.7803       11.7588        1.0153  ...   \n",
       "2     3895.0588     1536.1375     1084.5276      127.7205      245.1014  ...   \n",
       "3     1177.2241      873.9074      845.1423      229.8548      134.4694  ...   \n",
       "\n",
       "   GSM2701937  GSM2701938  GSM2701939  GSM2701940  GSM2701941  GSM2701942  \\\n",
       "0     17.4324     13.7006      9.8935     11.9023     11.3564      8.9213   \n",
       "1      2.0000      2.0443      1.0000      5.5140      7.6165      6.6037   \n",
       "2    165.9822    217.9993    114.9995    307.4970    272.9994    301.4232   \n",
       "3     96.6508     84.5137     70.8997     99.7864    145.2329    122.9726   \n",
       "\n",
       "   GSM2701943  GSM2701944  GSM2701945  GSM2701946  \n",
       "0     16.4894     13.9479     16.8066      7.7069  \n",
       "1      3.0000      7.0000      4.5765      1.0551  \n",
       "2    211.9566    227.9990    264.4285    256.9996  \n",
       "3     81.5081     79.0215     99.9146     77.1152  \n",
       "\n",
       "[4 rows x 1635 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = os.path.join('download', 'grein_count_matrix_human.pkl')\n",
    "expr_df = pd.read_pickle(file)\n",
    "\n",
    "print(expr_df.shape)\n",
    "expr_df.head(4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process gene expression matrix\n",
    "\n",
    "This involves updating Entrez gene ids, sorting and subsetting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(27990, 1635)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>GSM2667747_x</th>\n",
       "      <th>GSM2667748_x</th>\n",
       "      <th>GSM2667749_x</th>\n",
       "      <th>GSM2667750_x</th>\n",
       "      <th>GSM2667751_x</th>\n",
       "      <th>GSM2667752_x</th>\n",
       "      <th>GSM2667753_x</th>\n",
       "      <th>GSM2667754_x</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2701937</th>\n",
       "      <th>GSM2701938</th>\n",
       "      <th>GSM2701939</th>\n",
       "      <th>GSM2701940</th>\n",
       "      <th>GSM2701941</th>\n",
       "      <th>GSM2701942</th>\n",
       "      <th>GSM2701943</th>\n",
       "      <th>GSM2701944</th>\n",
       "      <th>GSM2701945</th>\n",
       "      <th>GSM2701946</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1766.0677</td>\n",
       "      <td>420.8206</td>\n",
       "      <td>300.8009</td>\n",
       "      <td>3142.8439</td>\n",
       "      <td>2207.8392</td>\n",
       "      <td>4367.5056</td>\n",
       "      <td>228.5759</td>\n",
       "      <td>650.3287</td>\n",
       "      <td>...</td>\n",
       "      <td>17.4324</td>\n",
       "      <td>13.7006</td>\n",
       "      <td>9.8935</td>\n",
       "      <td>11.9023</td>\n",
       "      <td>11.3564</td>\n",
       "      <td>8.9213</td>\n",
       "      <td>16.4894</td>\n",
       "      <td>13.9479</td>\n",
       "      <td>16.8066</td>\n",
       "      <td>7.7069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>43.9280</td>\n",
       "      <td>18.0091</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>115.8917</td>\n",
       "      <td>63.5936</td>\n",
       "      <td>39.7803</td>\n",
       "      <td>11.7588</td>\n",
       "      <td>1.0153</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0000</td>\n",
       "      <td>2.0443</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>5.5140</td>\n",
       "      <td>7.6165</td>\n",
       "      <td>6.6037</td>\n",
       "      <td>3.0000</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>4.5765</td>\n",
       "      <td>1.0551</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 1635 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      gene gene_symbol  GSM2667747_x  GSM2667748_x  \\\n",
       "sample_id                                                            \n",
       "0          ENSG00000000003      TSPAN6     1766.0677      420.8206   \n",
       "1          ENSG00000000005        TNMD       43.9280       18.0091   \n",
       "\n",
       "           GSM2667749_x  GSM2667750_x  GSM2667751_x  GSM2667752_x  \\\n",
       "sample_id                                                           \n",
       "0              300.8009     3142.8439     2207.8392     4367.5056   \n",
       "1                0.0000      115.8917       63.5936       39.7803   \n",
       "\n",
       "           GSM2667753_x  GSM2667754_x  ...  GSM2701937  GSM2701938  \\\n",
       "sample_id                              ...                           \n",
       "0              228.5759      650.3287  ...     17.4324     13.7006   \n",
       "1               11.7588        1.0153  ...      2.0000      2.0443   \n",
       "\n",
       "           GSM2701939  GSM2701940  GSM2701941  GSM2701942  GSM2701943  \\\n",
       "sample_id                                                               \n",
       "0              9.8935     11.9023     11.3564      8.9213     16.4894   \n",
       "1              1.0000      5.5140      7.6165      6.6037      3.0000   \n",
       "\n",
       "           GSM2701944  GSM2701945  GSM2701946  \n",
       "sample_id                                      \n",
       "0             13.9479     16.8066      7.7069  \n",
       "1              7.0000      4.5765      1.0551  \n",
       "\n",
       "[2 rows x 1635 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# expr_df = (expr_df\n",
    "#     .dropna(axis='rows')\n",
    "#     .reindex(probe_map_df.id)\n",
    "#     .rename(index=ensembl_to_entrez)\n",
    "#     .rename(index=old_to_new_entrez)\n",
    "#     .groupby(level=0).mean()\n",
    "#     .transpose()\n",
    "#     .sort_index(axis='rows')\n",
    "#     .sort_index(axis='columns')\n",
    "# )\n",
    "\n",
    "expr_df.index.rename('sample_id', inplace=True)\n",
    "\n",
    "print(expr_df.shape)\n",
    "expr_df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stratify Balanced Training and Testing Sets in TARGET Gene Expression\n",
    "\n",
    "Output training and testing gene expression datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#strat = pheno_df.set_index('sample_id').reindex(expr_df.index).primary_disease_code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cancertype_count_df = (\n",
    "#     pd.DataFrame(strat.value_counts()) #not using value_counts, what number should we use here? \n",
    "#     .reset_index()\n",
    "#     .rename({'index': 'cancertype', 'primary_disease_code': 'n ='}, axis='columns')\n",
    "# )\n",
    "\n",
    "# file = os.path.join('data', 'target_sample_counts.tsv') #change which file - do we have a file that works for this?\n",
    "# cancertype_count_df.to_csv(file, sep='\\t', index=False)\n",
    "\n",
    "# cancertype_count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df, test_df = train_test_split(expr_df,\n",
    "                                     test_size=0.1,\n",
    "                                     random_state=123) #if no stratify defined, should just randomize on its own"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(25191, 1635)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(2799, 1635)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(train_df.shape)\n",
    "test_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save train dataframe to file \n",
    "train_file = os.path.join('data', 'train_grein_processed.tsv.gz')\n",
    "train_df.to_csv(train_file, sep='\\t', compression='gzip', float_format='%.3g')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save test dataframe to file \n",
    "test_file = os.path.join('data', 'test_grein_processed.tsv.gz')\n",
    "test_df.to_csv(test_file, sep='\\t', compression='gzip', float_format='%.3g')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>GSM2667747_x</th>\n",
       "      <th>GSM2667748_x</th>\n",
       "      <th>GSM2667749_x</th>\n",
       "      <th>GSM2667750_x</th>\n",
       "      <th>GSM2667751_x</th>\n",
       "      <th>GSM2667752_x</th>\n",
       "      <th>GSM2667753_x</th>\n",
       "      <th>GSM2667754_x</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2701937</th>\n",
       "      <th>GSM2701938</th>\n",
       "      <th>GSM2701939</th>\n",
       "      <th>GSM2701940</th>\n",
       "      <th>GSM2701941</th>\n",
       "      <th>GSM2701942</th>\n",
       "      <th>GSM2701943</th>\n",
       "      <th>GSM2701944</th>\n",
       "      <th>GSM2701945</th>\n",
       "      <th>GSM2701946</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27744</th>\n",
       "      <td>ENSG00000283046</td>\n",
       "      <td>HTN1</td>\n",
       "      <td>125.4665</td>\n",
       "      <td>140.2465</td>\n",
       "      <td>17.6274</td>\n",
       "      <td>69.7745</td>\n",
       "      <td>73.7733</td>\n",
       "      <td>27.6203</td>\n",
       "      <td>98.1395</td>\n",
       "      <td>12.7103</td>\n",
       "      <td>...</td>\n",
       "      <td>13.4686</td>\n",
       "      <td>11.2211</td>\n",
       "      <td>3.8547</td>\n",
       "      <td>13.0525</td>\n",
       "      <td>4.3899</td>\n",
       "      <td>7.9562</td>\n",
       "      <td>13.4648</td>\n",
       "      <td>15.0952</td>\n",
       "      <td>12.8111</td>\n",
       "      <td>9.5805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9537</th>\n",
       "      <td>ENSG00000153037</td>\n",
       "      <td>SRP19</td>\n",
       "      <td>894.5558</td>\n",
       "      <td>357.1081</td>\n",
       "      <td>187.4946</td>\n",
       "      <td>1421.0479</td>\n",
       "      <td>456.0143</td>\n",
       "      <td>802.6670</td>\n",
       "      <td>124.8560</td>\n",
       "      <td>164.7368</td>\n",
       "      <td>...</td>\n",
       "      <td>246.7201</td>\n",
       "      <td>229.3536</td>\n",
       "      <td>137.9754</td>\n",
       "      <td>276.6840</td>\n",
       "      <td>283.5146</td>\n",
       "      <td>315.5401</td>\n",
       "      <td>315.4701</td>\n",
       "      <td>327.8115</td>\n",
       "      <td>311.3958</td>\n",
       "      <td>287.1789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9769</th>\n",
       "      <td>ENSG00000155115</td>\n",
       "      <td>GTF3C6</td>\n",
       "      <td>851.6143</td>\n",
       "      <td>187.0203</td>\n",
       "      <td>243.4404</td>\n",
       "      <td>2046.5828</td>\n",
       "      <td>476.0532</td>\n",
       "      <td>1977.8905</td>\n",
       "      <td>94.7822</td>\n",
       "      <td>881.8821</td>\n",
       "      <td>...</td>\n",
       "      <td>271.8578</td>\n",
       "      <td>257.8892</td>\n",
       "      <td>179.5595</td>\n",
       "      <td>371.9447</td>\n",
       "      <td>309.9999</td>\n",
       "      <td>370.9075</td>\n",
       "      <td>199.8807</td>\n",
       "      <td>206.8743</td>\n",
       "      <td>194.0005</td>\n",
       "      <td>199.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15239</th>\n",
       "      <td>ENSG00000184083</td>\n",
       "      <td>FAM120C</td>\n",
       "      <td>632.3881</td>\n",
       "      <td>285.1058</td>\n",
       "      <td>122.8606</td>\n",
       "      <td>832.8702</td>\n",
       "      <td>957.4550</td>\n",
       "      <td>1495.5920</td>\n",
       "      <td>184.7934</td>\n",
       "      <td>91.7322</td>\n",
       "      <td>...</td>\n",
       "      <td>75.4221</td>\n",
       "      <td>50.0695</td>\n",
       "      <td>31.4341</td>\n",
       "      <td>54.8382</td>\n",
       "      <td>67.9711</td>\n",
       "      <td>66.0159</td>\n",
       "      <td>73.3668</td>\n",
       "      <td>45.4544</td>\n",
       "      <td>63.2615</td>\n",
       "      <td>41.2241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>ENSG00000008196</td>\n",
       "      <td>TFAP2B</td>\n",
       "      <td>23.2873</td>\n",
       "      <td>2.0018</td>\n",
       "      <td>19.1983</td>\n",
       "      <td>54.7994</td>\n",
       "      <td>19.0940</td>\n",
       "      <td>974.5680</td>\n",
       "      <td>88.7088</td>\n",
       "      <td>88.8378</td>\n",
       "      <td>...</td>\n",
       "      <td>20.4597</td>\n",
       "      <td>24.2683</td>\n",
       "      <td>14.5826</td>\n",
       "      <td>31.6125</td>\n",
       "      <td>12.5990</td>\n",
       "      <td>10.9160</td>\n",
       "      <td>16.7363</td>\n",
       "      <td>16.5086</td>\n",
       "      <td>22.4091</td>\n",
       "      <td>15.6715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15377</th>\n",
       "      <td>ENSG00000184730</td>\n",
       "      <td>APOBR</td>\n",
       "      <td>6.3818</td>\n",
       "      <td>9.1460</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>11.3041</td>\n",
       "      <td>10.1788</td>\n",
       "      <td>9.9309</td>\n",
       "      <td>8.7824</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>4218.6053</td>\n",
       "      <td>4018.6666</td>\n",
       "      <td>1948.5265</td>\n",
       "      <td>3348.8270</td>\n",
       "      <td>3218.9786</td>\n",
       "      <td>3721.1712</td>\n",
       "      <td>3864.2544</td>\n",
       "      <td>3813.3692</td>\n",
       "      <td>3034.6335</td>\n",
       "      <td>2921.8278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21602</th>\n",
       "      <td>ENSG00000235346</td>\n",
       "      <td>HLA-G</td>\n",
       "      <td>22.1629</td>\n",
       "      <td>2.8791</td>\n",
       "      <td>12.0961</td>\n",
       "      <td>68.9133</td>\n",
       "      <td>108.9510</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2.0205</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6717</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.0915</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2.8872</td>\n",
       "      <td>4.0845</td>\n",
       "      <td>1.1946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17730</th>\n",
       "      <td>ENSG00000204291</td>\n",
       "      <td>COL15A1</td>\n",
       "      <td>47.6025</td>\n",
       "      <td>26.1180</td>\n",
       "      <td>7.0183</td>\n",
       "      <td>141.7424</td>\n",
       "      <td>56.5391</td>\n",
       "      <td>392.2496</td>\n",
       "      <td>27.0797</td>\n",
       "      <td>59.5957</td>\n",
       "      <td>...</td>\n",
       "      <td>52.6309</td>\n",
       "      <td>26.4790</td>\n",
       "      <td>24.0000</td>\n",
       "      <td>26.1559</td>\n",
       "      <td>32.5506</td>\n",
       "      <td>26.8942</td>\n",
       "      <td>42.6889</td>\n",
       "      <td>45.5209</td>\n",
       "      <td>32.9535</td>\n",
       "      <td>23.5736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15725</th>\n",
       "      <td>ENSG00000186191</td>\n",
       "      <td>BPIFB4</td>\n",
       "      <td>3.9720</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2.0215</td>\n",
       "      <td>5.0149</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>18.0344</td>\n",
       "      <td>24.0000</td>\n",
       "      <td>14.6045</td>\n",
       "      <td>24.0000</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>5.0000</td>\n",
       "      <td>43.8003</td>\n",
       "      <td>26.1682</td>\n",
       "      <td>24.0000</td>\n",
       "      <td>14.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19966</th>\n",
       "      <td>ENSG00000226023</td>\n",
       "      <td>CT47A6</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25191 rows Ã— 1635 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      gene gene_symbol  GSM2667747_x  GSM2667748_x  \\\n",
       "sample_id                                                            \n",
       "27744      ENSG00000283046        HTN1      125.4665      140.2465   \n",
       "9537       ENSG00000153037       SRP19      894.5558      357.1081   \n",
       "9769       ENSG00000155115      GTF3C6      851.6143      187.0203   \n",
       "15239      ENSG00000184083     FAM120C      632.3881      285.1058   \n",
       "203        ENSG00000008196      TFAP2B       23.2873        2.0018   \n",
       "...                    ...         ...           ...           ...   \n",
       "15377      ENSG00000184730       APOBR        6.3818        9.1460   \n",
       "21602      ENSG00000235346       HLA-G       22.1629        2.8791   \n",
       "17730      ENSG00000204291     COL15A1       47.6025       26.1180   \n",
       "15725      ENSG00000186191      BPIFB4        3.9720        0.0000   \n",
       "19966      ENSG00000226023      CT47A6        0.0000        0.0000   \n",
       "\n",
       "           GSM2667749_x  GSM2667750_x  GSM2667751_x  GSM2667752_x  \\\n",
       "sample_id                                                           \n",
       "27744           17.6274       69.7745       73.7733       27.6203   \n",
       "9537           187.4946     1421.0479      456.0143      802.6670   \n",
       "9769           243.4404     2046.5828      476.0532     1977.8905   \n",
       "15239          122.8606      832.8702      957.4550     1495.5920   \n",
       "203             19.1983       54.7994       19.0940      974.5680   \n",
       "...                 ...           ...           ...           ...   \n",
       "15377            0.0000       11.3041       10.1788        9.9309   \n",
       "21602           12.0961       68.9133      108.9510        0.0000   \n",
       "17730            7.0183      141.7424       56.5391      392.2496   \n",
       "15725            0.0000        2.0215        5.0149        0.0000   \n",
       "19966            0.0000        0.0000        0.0000        0.0000   \n",
       "\n",
       "           GSM2667753_x  GSM2667754_x  ...  GSM2701937  GSM2701938  \\\n",
       "sample_id                              ...                           \n",
       "27744           98.1395       12.7103  ...     13.4686     11.2211   \n",
       "9537           124.8560      164.7368  ...    246.7201    229.3536   \n",
       "9769            94.7822      881.8821  ...    271.8578    257.8892   \n",
       "15239          184.7934       91.7322  ...     75.4221     50.0695   \n",
       "203             88.7088       88.8378  ...     20.4597     24.2683   \n",
       "...                 ...           ...  ...         ...         ...   \n",
       "15377            8.7824        0.0000  ...   4218.6053   4018.6666   \n",
       "21602            0.0000        2.0205  ...      0.6717      0.0000   \n",
       "17730           27.0797       59.5957  ...     52.6309     26.4790   \n",
       "15725            0.0000        0.0000  ...     18.0344     24.0000   \n",
       "19966            0.0000        0.0000  ...      0.0000      0.0000   \n",
       "\n",
       "           GSM2701939  GSM2701940  GSM2701941  GSM2701942  GSM2701943  \\\n",
       "sample_id                                                               \n",
       "27744          3.8547     13.0525      4.3899      7.9562     13.4648   \n",
       "9537         137.9754    276.6840    283.5146    315.5401    315.4701   \n",
       "9769         179.5595    371.9447    309.9999    370.9075    199.8807   \n",
       "15239         31.4341     54.8382     67.9711     66.0159     73.3668   \n",
       "203           14.5826     31.6125     12.5990     10.9160     16.7363   \n",
       "...               ...         ...         ...         ...         ...   \n",
       "15377       1948.5265   3348.8270   3218.9786   3721.1712   3864.2544   \n",
       "21602          1.0915      0.0000      0.0000      0.0000      0.0000   \n",
       "17730         24.0000     26.1559     32.5506     26.8942     42.6889   \n",
       "15725         14.6045     24.0000      7.0000      5.0000     43.8003   \n",
       "19966          0.0000      0.0000      0.0000      0.0000      0.0000   \n",
       "\n",
       "           GSM2701944  GSM2701945  GSM2701946  \n",
       "sample_id                                      \n",
       "27744         15.0952     12.8111      9.5805  \n",
       "9537         327.8115    311.3958    287.1789  \n",
       "9769         206.8743    194.0005    199.0000  \n",
       "15239         45.4544     63.2615     41.2241  \n",
       "203           16.5086     22.4091     15.6715  \n",
       "...               ...         ...         ...  \n",
       "15377       3813.3692   3034.6335   2921.8278  \n",
       "21602          2.8872      4.0845      1.1946  \n",
       "17730         45.5209     32.9535     23.5736  \n",
       "15725         26.1682     24.0000     14.0000  \n",
       "19966          0.0000      0.0000      0.0002  \n",
       "\n",
       "[25191 rows x 1635 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sort genes based on median absolute deviation and output to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'method' object is not subscriptable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[23], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m train_df_mad \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mabs\u001b[39m(train_df \u001b[38;5;241m-\u001b[39m \u001b[43mtrain_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdrop\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msample_id\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgene\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgene_symbol\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mmedian())\u001b[38;5;241m.\u001b[39mmedian()\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28mprint\u001b[39m(train_df_mad)\n",
      "\u001b[0;31mTypeError\u001b[0m: 'method' object is not subscriptable"
     ]
    }
   ],
   "source": [
    "train_df_mad = abs(train_df - train_df.drop(['sample_id', 'gene', 'gene_symbol'], axis = 1).median()).median()\n",
    "print(train_df_mad)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'median_abs_deviation'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_1952/3658105903.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Determine most variably expressed genes and subset\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mmad_genes_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtrain_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmedian_abs_deviation\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mascending\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mmad_genes_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'gene_id'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'median_abs_deviation'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mfile\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'data'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'target_mad_genes.tsv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/micromamba/envs/biobombe/lib/python3.12/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6295\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6296\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6297\u001b[0m         \u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6298\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6299\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'median_abs_deviation'"
     ]
    }
   ],
   "source": [
    "# Determine most variably expressed genes and subset\n",
    "train_df_mad = abs(train_df - train_df.mean()).mean()\n",
    "mad_genes_df = pd.DataFrame(train_df.median_abs_deviation(axis=0).sort_values(ascending=False)).reset_index()\n",
    "mad_genes_df.columns = ['gene_id', 'median_abs_deviation']\n",
    "\n",
    "file = os.path.join('data', 'target_mad_genes.tsv')\n",
    "mad_genes_df.to_csv(file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
