{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process GREIN Rat Data\n",
    "\n",
    "Retrieve the downloaded expression data, update gene identifiers to entrez, and curate sample IDs. The script will also identify a balanced hold-out test set to compare projection performance into learned latent spaces across algorithms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import random\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(1234)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Phenotype Information: skipping until phenotype data found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path = os.path.join('download', 'TARGET_phenotype.gz')\n",
    "# pheno_df = pd.read_table(path)\n",
    "\n",
    "# print(pheno_df.shape)\n",
    "# pheno_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Probe Mapping Info \n",
    "(where chromosomes start and end, saved in a file in downloads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path = os.path.join('download', 'gencode.v23.annotation.gene.probemap')\n",
    "# probe_map_df = pd.read_table(path)\n",
    "\n",
    "# # Inner merge gene df to get ensembl to entrez mapping\n",
    "# probe_map_df = probe_map_df.merge(gene_df, how='inner', left_on='gene', right_on='symbol')\n",
    "\n",
    "# # Mapping to rename gene expression index\n",
    "# ensembl_to_entrez = dict(zip(probe_map_df.id, probe_map_df.entrez_gene_id))\n",
    "\n",
    "# print(probe_map_df.shape)\n",
    "# probe_map_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Gene Expression Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18171, 39)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>GSM2668015</th>\n",
       "      <th>GSM2668016</th>\n",
       "      <th>GSM2668017</th>\n",
       "      <th>GSM2668018</th>\n",
       "      <th>GSM2668019</th>\n",
       "      <th>GSM2668020</th>\n",
       "      <th>GSM2668021</th>\n",
       "      <th>GSM2668022</th>\n",
       "      <th>GSM2674552</th>\n",
       "      <th>GSM2674553</th>\n",
       "      <th>GSM2674554</th>\n",
       "      <th>GSM2674555</th>\n",
       "      <th>GSM2674556</th>\n",
       "      <th>GSM2674557</th>\n",
       "      <th>GSM2674558</th>\n",
       "      <th>GSM2674559</th>\n",
       "      <th>GSM2674560</th>\n",
       "      <th>GSM2674561</th>\n",
       "      <th>GSM2674562</th>\n",
       "      <th>GSM2674563</th>\n",
       "      <th>GSM2674564</th>\n",
       "      <th>GSM2674565</th>\n",
       "      <th>GSM2674566</th>\n",
       "      <th>GSM2674567</th>\n",
       "      <th>GSM2683306</th>\n",
       "      <th>GSM2683307</th>\n",
       "      <th>GSM2683308</th>\n",
       "      <th>GSM2683309</th>\n",
       "      <th>GSM2683310</th>\n",
       "      <th>GSM2683311</th>\n",
       "      <th>GSM2683312</th>\n",
       "      <th>GSM2699989</th>\n",
       "      <th>GSM2699990</th>\n",
       "      <th>GSM2699991</th>\n",
       "      <th>GSM2699992</th>\n",
       "      <th>GSM2699993</th>\n",
       "      <th>GSM2699994</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSRNOG00000000007</td>\n",
       "      <td>Gad1</td>\n",
       "      <td>5140.2551</td>\n",
       "      <td>2873.5173</td>\n",
       "      <td>3494.3567</td>\n",
       "      <td>2488.5614</td>\n",
       "      <td>2180.0086</td>\n",
       "      <td>4054.4603</td>\n",
       "      <td>3249.2422</td>\n",
       "      <td>2984.1460</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.5118</td>\n",
       "      <td>27.5834</td>\n",
       "      <td>70.6123</td>\n",
       "      <td>55.3001</td>\n",
       "      <td>60.6636</td>\n",
       "      <td>64.8627</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSRNOG00000000008</td>\n",
       "      <td>Alx4</td>\n",
       "      <td>88.3117</td>\n",
       "      <td>54.3955</td>\n",
       "      <td>68.8133</td>\n",
       "      <td>54.5058</td>\n",
       "      <td>61.3839</td>\n",
       "      <td>84.8890</td>\n",
       "      <td>92.7423</td>\n",
       "      <td>62.1657</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.2676</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>101.9639</td>\n",
       "      <td>77.8544</td>\n",
       "      <td>77.2475</td>\n",
       "      <td>67.0283</td>\n",
       "      <td>61.0076</td>\n",
       "      <td>92.9832</td>\n",
       "      <td>75.9176</td>\n",
       "      <td>10.0072</td>\n",
       "      <td>10.2904</td>\n",
       "      <td>12.9965</td>\n",
       "      <td>5.2917</td>\n",
       "      <td>2.9983</td>\n",
       "      <td>2.9982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSRNOG00000000009</td>\n",
       "      <td>Tmco5b</td>\n",
       "      <td>71.9091</td>\n",
       "      <td>47.7813</td>\n",
       "      <td>50.4332</td>\n",
       "      <td>41.1147</td>\n",
       "      <td>40.6470</td>\n",
       "      <td>56.4535</td>\n",
       "      <td>48.4759</td>\n",
       "      <td>47.5353</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.9727</td>\n",
       "      <td>42.2429</td>\n",
       "      <td>54.0743</td>\n",
       "      <td>83.0249</td>\n",
       "      <td>56.5682</td>\n",
       "      <td>75.6148</td>\n",
       "      <td>44.3473</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.9925</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSRNOG00000000010</td>\n",
       "      <td>Cbln1</td>\n",
       "      <td>366.8392</td>\n",
       "      <td>236.4766</td>\n",
       "      <td>388.1470</td>\n",
       "      <td>212.8647</td>\n",
       "      <td>218.2449</td>\n",
       "      <td>392.4068</td>\n",
       "      <td>325.6375</td>\n",
       "      <td>231.4440</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>126.6326</td>\n",
       "      <td>97.5128</td>\n",
       "      <td>89.0460</td>\n",
       "      <td>141.5510</td>\n",
       "      <td>100.1295</td>\n",
       "      <td>143.5802</td>\n",
       "      <td>87.1007</td>\n",
       "      <td>80.7889</td>\n",
       "      <td>95.8700</td>\n",
       "      <td>88.7592</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9978</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene gene_symbol  GSM2668015  GSM2668016  GSM2668017  \\\n",
       "0  ENSRNOG00000000007        Gad1   5140.2551   2873.5173   3494.3567   \n",
       "1  ENSRNOG00000000008        Alx4     88.3117     54.3955     68.8133   \n",
       "2  ENSRNOG00000000009      Tmco5b     71.9091     47.7813     50.4332   \n",
       "3  ENSRNOG00000000010       Cbln1    366.8392    236.4766    388.1470   \n",
       "\n",
       "   GSM2668018  GSM2668019  GSM2668020  GSM2668021  GSM2668022  GSM2674552  \\\n",
       "0   2488.5614   2180.0086   4054.4603   3249.2422   2984.1460         0.0   \n",
       "1     54.5058     61.3839     84.8890     92.7423     62.1657         0.0   \n",
       "2     41.1147     40.6470     56.4535     48.4759     47.5353         0.0   \n",
       "3    212.8647    218.2449    392.4068    325.6375    231.4440         0.0   \n",
       "\n",
       "   GSM2674553  GSM2674554  GSM2674555  GSM2674556  GSM2674557  GSM2674558  \\\n",
       "0         0.0         0.0         0.0         0.0         0.0      0.0000   \n",
       "1         0.0         0.0         2.0         0.0         2.0      1.2676   \n",
       "2         0.0         0.0         0.0         0.0         0.0      0.0000   \n",
       "3         0.0         0.0         0.0         0.0         0.0      0.0000   \n",
       "\n",
       "   GSM2674559  GSM2674560  GSM2674561  GSM2674562  GSM2674563  GSM2674564  \\\n",
       "0         0.0         0.0         0.0         0.0         0.0      0.0000   \n",
       "1         1.0         0.0         1.0         0.0         2.0      1.0000   \n",
       "2         0.0         0.0         0.0         0.0         0.0      0.9999   \n",
       "3         0.0         0.0         0.0         0.0         0.0      0.0000   \n",
       "\n",
       "   GSM2674565  GSM2674566  GSM2674567  GSM2683306  GSM2683307  GSM2683308  \\\n",
       "0         0.0         0.0         0.0     52.5118     27.5834     70.6123   \n",
       "1         0.0         0.0         2.0    101.9639     77.8544     77.2475   \n",
       "2         0.0         0.0         0.0     76.9727     42.2429     54.0743   \n",
       "3         0.0         0.0         0.0    126.6326     97.5128     89.0460   \n",
       "\n",
       "   GSM2683309  GSM2683310  GSM2683311  GSM2683312  GSM2699989  GSM2699990  \\\n",
       "0     55.3001     60.6636     64.8627      0.0000      0.0000      0.0000   \n",
       "1     67.0283     61.0076     92.9832     75.9176     10.0072     10.2904   \n",
       "2     83.0249     56.5682     75.6148     44.3473      0.0000      0.0000   \n",
       "3    141.5510    100.1295    143.5802     87.1007     80.7889     95.8700   \n",
       "\n",
       "   GSM2699991  GSM2699992  GSM2699993  GSM2699994  \n",
       "0      0.0000      0.0000      0.0000      0.0000  \n",
       "1     12.9965      5.2917      2.9983      2.9982  \n",
       "2      1.9925      0.0000      0.0000      0.0000  \n",
       "3     88.7592      0.0000      0.9978      0.0000  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reading in rat gene expression data\n",
    "\n",
    "file = os.path.join('download', 'grein_count_matrix_rat.pkl')\n",
    "expr_df = pd.read_pickle(file)\n",
    "\n",
    "print(expr_df.shape)\n",
    "expr_df.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean up x and y from expr_df\n",
    "\n",
    "# create a list of cleaned columns\n",
    "cleaned_column = ['gene', 'gene_symbol']\n",
    "for column in expr_df.drop(['gene', 'gene_symbol'], axis = 1).columns:\n",
    "    cleaned_column.append(column.split(\"_\")[0])\n",
    "\n",
    "# set colums to cleaned column\n",
    "expr_df.columns = cleaned_column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>GSM2668015</th>\n",
       "      <th>GSM2668016</th>\n",
       "      <th>GSM2668017</th>\n",
       "      <th>GSM2668018</th>\n",
       "      <th>GSM2668019</th>\n",
       "      <th>GSM2668020</th>\n",
       "      <th>GSM2668021</th>\n",
       "      <th>GSM2668022</th>\n",
       "      <th>GSM2674552</th>\n",
       "      <th>GSM2674553</th>\n",
       "      <th>GSM2674554</th>\n",
       "      <th>GSM2674555</th>\n",
       "      <th>GSM2674556</th>\n",
       "      <th>GSM2674557</th>\n",
       "      <th>GSM2674558</th>\n",
       "      <th>GSM2674559</th>\n",
       "      <th>GSM2674560</th>\n",
       "      <th>GSM2674561</th>\n",
       "      <th>GSM2674562</th>\n",
       "      <th>GSM2674563</th>\n",
       "      <th>GSM2674564</th>\n",
       "      <th>GSM2674565</th>\n",
       "      <th>GSM2674566</th>\n",
       "      <th>GSM2674567</th>\n",
       "      <th>GSM2683306</th>\n",
       "      <th>GSM2683307</th>\n",
       "      <th>GSM2683308</th>\n",
       "      <th>GSM2683309</th>\n",
       "      <th>GSM2683310</th>\n",
       "      <th>GSM2683311</th>\n",
       "      <th>GSM2683312</th>\n",
       "      <th>GSM2699989</th>\n",
       "      <th>GSM2699990</th>\n",
       "      <th>GSM2699991</th>\n",
       "      <th>GSM2699992</th>\n",
       "      <th>GSM2699993</th>\n",
       "      <th>GSM2699994</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSRNOG00000000007</td>\n",
       "      <td>Gad1</td>\n",
       "      <td>5140.2551</td>\n",
       "      <td>2873.5173</td>\n",
       "      <td>3494.3567</td>\n",
       "      <td>2488.5614</td>\n",
       "      <td>2180.0086</td>\n",
       "      <td>4054.4603</td>\n",
       "      <td>3249.2422</td>\n",
       "      <td>2984.1460</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.5118</td>\n",
       "      <td>27.5834</td>\n",
       "      <td>70.6123</td>\n",
       "      <td>55.3001</td>\n",
       "      <td>60.6636</td>\n",
       "      <td>64.8627</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSRNOG00000000008</td>\n",
       "      <td>Alx4</td>\n",
       "      <td>88.3117</td>\n",
       "      <td>54.3955</td>\n",
       "      <td>68.8133</td>\n",
       "      <td>54.5058</td>\n",
       "      <td>61.3839</td>\n",
       "      <td>84.8890</td>\n",
       "      <td>92.7423</td>\n",
       "      <td>62.1657</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.2676</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>101.9639</td>\n",
       "      <td>77.8544</td>\n",
       "      <td>77.2475</td>\n",
       "      <td>67.0283</td>\n",
       "      <td>61.0076</td>\n",
       "      <td>92.9832</td>\n",
       "      <td>75.9176</td>\n",
       "      <td>10.0072</td>\n",
       "      <td>10.2904</td>\n",
       "      <td>12.9965</td>\n",
       "      <td>5.2917</td>\n",
       "      <td>2.9983</td>\n",
       "      <td>2.9982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSRNOG00000000009</td>\n",
       "      <td>Tmco5b</td>\n",
       "      <td>71.9091</td>\n",
       "      <td>47.7813</td>\n",
       "      <td>50.4332</td>\n",
       "      <td>41.1147</td>\n",
       "      <td>40.6470</td>\n",
       "      <td>56.4535</td>\n",
       "      <td>48.4759</td>\n",
       "      <td>47.5353</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.9727</td>\n",
       "      <td>42.2429</td>\n",
       "      <td>54.0743</td>\n",
       "      <td>83.0249</td>\n",
       "      <td>56.5682</td>\n",
       "      <td>75.6148</td>\n",
       "      <td>44.3473</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.9925</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSRNOG00000000010</td>\n",
       "      <td>Cbln1</td>\n",
       "      <td>366.8392</td>\n",
       "      <td>236.4766</td>\n",
       "      <td>388.1470</td>\n",
       "      <td>212.8647</td>\n",
       "      <td>218.2449</td>\n",
       "      <td>392.4068</td>\n",
       "      <td>325.6375</td>\n",
       "      <td>231.4440</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>126.6326</td>\n",
       "      <td>97.5128</td>\n",
       "      <td>89.0460</td>\n",
       "      <td>141.5510</td>\n",
       "      <td>100.1295</td>\n",
       "      <td>143.5802</td>\n",
       "      <td>87.1007</td>\n",
       "      <td>80.7889</td>\n",
       "      <td>95.8700</td>\n",
       "      <td>88.7592</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9978</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSRNOG00000000012</td>\n",
       "      <td>Tcf15</td>\n",
       "      <td>164.0565</td>\n",
       "      <td>86.1982</td>\n",
       "      <td>86.2113</td>\n",
       "      <td>86.0566</td>\n",
       "      <td>100.7740</td>\n",
       "      <td>145.9362</td>\n",
       "      <td>98.5538</td>\n",
       "      <td>97.6332</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.7998</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.4823</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2.4423</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.8007</td>\n",
       "      <td>33.2070</td>\n",
       "      <td>33.8593</td>\n",
       "      <td>47.3560</td>\n",
       "      <td>26.0843</td>\n",
       "      <td>30.6765</td>\n",
       "      <td>43.2454</td>\n",
       "      <td>9.9237</td>\n",
       "      <td>13.7869</td>\n",
       "      <td>6.8718</td>\n",
       "      <td>3.3902</td>\n",
       "      <td>5.8108</td>\n",
       "      <td>10.7463</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene gene_symbol  GSM2668015  GSM2668016  GSM2668017  \\\n",
       "0  ENSRNOG00000000007        Gad1   5140.2551   2873.5173   3494.3567   \n",
       "1  ENSRNOG00000000008        Alx4     88.3117     54.3955     68.8133   \n",
       "2  ENSRNOG00000000009      Tmco5b     71.9091     47.7813     50.4332   \n",
       "3  ENSRNOG00000000010       Cbln1    366.8392    236.4766    388.1470   \n",
       "4  ENSRNOG00000000012       Tcf15    164.0565     86.1982     86.2113   \n",
       "\n",
       "   GSM2668018  GSM2668019  GSM2668020  GSM2668021  GSM2668022  GSM2674552  \\\n",
       "0   2488.5614   2180.0086   4054.4603   3249.2422   2984.1460         0.0   \n",
       "1     54.5058     61.3839     84.8890     92.7423     62.1657         0.0   \n",
       "2     41.1147     40.6470     56.4535     48.4759     47.5353         0.0   \n",
       "3    212.8647    218.2449    392.4068    325.6375    231.4440         0.0   \n",
       "4     86.0566    100.7740    145.9362     98.5538     97.6332         0.0   \n",
       "\n",
       "   GSM2674553  GSM2674554  GSM2674555  GSM2674556  GSM2674557  GSM2674558  \\\n",
       "0      0.0000         0.0      0.0000         0.0         0.0      0.0000   \n",
       "1      0.0000         0.0      2.0000         0.0         2.0      1.2676   \n",
       "2      0.0000         0.0      0.0000         0.0         0.0      0.0000   \n",
       "3      0.0000         0.0      0.0000         0.0         0.0      0.0000   \n",
       "4      1.7998         0.0      1.4823         0.0         0.0      0.0000   \n",
       "\n",
       "   GSM2674559  GSM2674560  GSM2674561  GSM2674562  GSM2674563  GSM2674564  \\\n",
       "0         0.0         0.0         0.0         0.0         0.0      0.0000   \n",
       "1         1.0         0.0         1.0         0.0         2.0      1.0000   \n",
       "2         0.0         0.0         0.0         0.0         0.0      0.9999   \n",
       "3         0.0         0.0         0.0         0.0         0.0      0.0000   \n",
       "4         0.0         0.0         0.0         0.0         0.0      0.0000   \n",
       "\n",
       "   GSM2674565  GSM2674566  GSM2674567  GSM2683306  GSM2683307  GSM2683308  \\\n",
       "0      0.0000         0.0         0.0     52.5118     27.5834     70.6123   \n",
       "1      0.0000         0.0         2.0    101.9639     77.8544     77.2475   \n",
       "2      0.0000         0.0         0.0     76.9727     42.2429     54.0743   \n",
       "3      0.0000         0.0         0.0    126.6326     97.5128     89.0460   \n",
       "4      2.4423         0.0         0.0     31.8007     33.2070     33.8593   \n",
       "\n",
       "   GSM2683309  GSM2683310  GSM2683311  GSM2683312  GSM2699989  GSM2699990  \\\n",
       "0     55.3001     60.6636     64.8627      0.0000      0.0000      0.0000   \n",
       "1     67.0283     61.0076     92.9832     75.9176     10.0072     10.2904   \n",
       "2     83.0249     56.5682     75.6148     44.3473      0.0000      0.0000   \n",
       "3    141.5510    100.1295    143.5802     87.1007     80.7889     95.8700   \n",
       "4     47.3560     26.0843     30.6765     43.2454      9.9237     13.7869   \n",
       "\n",
       "   GSM2699991  GSM2699992  GSM2699993  GSM2699994  \n",
       "0      0.0000      0.0000      0.0000      0.0000  \n",
       "1     12.9965      5.2917      2.9983      2.9982  \n",
       "2      1.9925      0.0000      0.0000      0.0000  \n",
       "3     88.7592      0.0000      0.9978      0.0000  \n",
       "4      6.8718      3.3902      5.8108     10.7463  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check that it worked\n",
    "\n",
    "pd.set_option('display.max_columns', None) \n",
    "expr_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process gene expression matrix\n",
    "\n",
    "This involves updating Entrez gene ids, sorting and subsetting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18171, 39)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>GSM2668015</th>\n",
       "      <th>GSM2668016</th>\n",
       "      <th>GSM2668017</th>\n",
       "      <th>GSM2668018</th>\n",
       "      <th>GSM2668019</th>\n",
       "      <th>GSM2668020</th>\n",
       "      <th>GSM2668021</th>\n",
       "      <th>GSM2668022</th>\n",
       "      <th>GSM2674552</th>\n",
       "      <th>GSM2674553</th>\n",
       "      <th>GSM2674554</th>\n",
       "      <th>GSM2674555</th>\n",
       "      <th>GSM2674556</th>\n",
       "      <th>GSM2674557</th>\n",
       "      <th>GSM2674558</th>\n",
       "      <th>GSM2674559</th>\n",
       "      <th>GSM2674560</th>\n",
       "      <th>GSM2674561</th>\n",
       "      <th>GSM2674562</th>\n",
       "      <th>GSM2674563</th>\n",
       "      <th>GSM2674564</th>\n",
       "      <th>GSM2674565</th>\n",
       "      <th>GSM2674566</th>\n",
       "      <th>GSM2674567</th>\n",
       "      <th>GSM2683306</th>\n",
       "      <th>GSM2683307</th>\n",
       "      <th>GSM2683308</th>\n",
       "      <th>GSM2683309</th>\n",
       "      <th>GSM2683310</th>\n",
       "      <th>GSM2683311</th>\n",
       "      <th>GSM2683312</th>\n",
       "      <th>GSM2699989</th>\n",
       "      <th>GSM2699990</th>\n",
       "      <th>GSM2699991</th>\n",
       "      <th>GSM2699992</th>\n",
       "      <th>GSM2699993</th>\n",
       "      <th>GSM2699994</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSRNOG00000000007</td>\n",
       "      <td>Gad1</td>\n",
       "      <td>5140.2551</td>\n",
       "      <td>2873.5173</td>\n",
       "      <td>3494.3567</td>\n",
       "      <td>2488.5614</td>\n",
       "      <td>2180.0086</td>\n",
       "      <td>4054.4603</td>\n",
       "      <td>3249.2422</td>\n",
       "      <td>2984.1460</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.5118</td>\n",
       "      <td>27.5834</td>\n",
       "      <td>70.6123</td>\n",
       "      <td>55.3001</td>\n",
       "      <td>60.6636</td>\n",
       "      <td>64.8627</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSRNOG00000000008</td>\n",
       "      <td>Alx4</td>\n",
       "      <td>88.3117</td>\n",
       "      <td>54.3955</td>\n",
       "      <td>68.8133</td>\n",
       "      <td>54.5058</td>\n",
       "      <td>61.3839</td>\n",
       "      <td>84.8890</td>\n",
       "      <td>92.7423</td>\n",
       "      <td>62.1657</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.2676</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>101.9639</td>\n",
       "      <td>77.8544</td>\n",
       "      <td>77.2475</td>\n",
       "      <td>67.0283</td>\n",
       "      <td>61.0076</td>\n",
       "      <td>92.9832</td>\n",
       "      <td>75.9176</td>\n",
       "      <td>10.0072</td>\n",
       "      <td>10.2904</td>\n",
       "      <td>12.9965</td>\n",
       "      <td>5.2917</td>\n",
       "      <td>2.9983</td>\n",
       "      <td>2.9982</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         gene gene_symbol  GSM2668015  GSM2668016  GSM2668017  \\\n",
       "sample_id                                                                       \n",
       "0          ENSRNOG00000000007        Gad1   5140.2551   2873.5173   3494.3567   \n",
       "1          ENSRNOG00000000008        Alx4     88.3117     54.3955     68.8133   \n",
       "\n",
       "           GSM2668018  GSM2668019  GSM2668020  GSM2668021  GSM2668022  \\\n",
       "sample_id                                                               \n",
       "0           2488.5614   2180.0086   4054.4603   3249.2422   2984.1460   \n",
       "1             54.5058     61.3839     84.8890     92.7423     62.1657   \n",
       "\n",
       "           GSM2674552  GSM2674553  GSM2674554  GSM2674555  GSM2674556  \\\n",
       "sample_id                                                               \n",
       "0                 0.0         0.0         0.0         0.0         0.0   \n",
       "1                 0.0         0.0         0.0         2.0         0.0   \n",
       "\n",
       "           GSM2674557  GSM2674558  GSM2674559  GSM2674560  GSM2674561  \\\n",
       "sample_id                                                               \n",
       "0                 0.0      0.0000         0.0         0.0         0.0   \n",
       "1                 2.0      1.2676         1.0         0.0         1.0   \n",
       "\n",
       "           GSM2674562  GSM2674563  GSM2674564  GSM2674565  GSM2674566  \\\n",
       "sample_id                                                               \n",
       "0                 0.0         0.0         0.0         0.0         0.0   \n",
       "1                 0.0         2.0         1.0         0.0         0.0   \n",
       "\n",
       "           GSM2674567  GSM2683306  GSM2683307  GSM2683308  GSM2683309  \\\n",
       "sample_id                                                               \n",
       "0                 0.0     52.5118     27.5834     70.6123     55.3001   \n",
       "1                 2.0    101.9639     77.8544     77.2475     67.0283   \n",
       "\n",
       "           GSM2683310  GSM2683311  GSM2683312  GSM2699989  GSM2699990  \\\n",
       "sample_id                                                               \n",
       "0             60.6636     64.8627      0.0000      0.0000      0.0000   \n",
       "1             61.0076     92.9832     75.9176     10.0072     10.2904   \n",
       "\n",
       "           GSM2699991  GSM2699992  GSM2699993  GSM2699994  \n",
       "sample_id                                                  \n",
       "0              0.0000      0.0000      0.0000      0.0000  \n",
       "1             12.9965      5.2917      2.9983      2.9982  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# expr_df = (expr_df\n",
    "#     .dropna(axis='rows')\n",
    "#     .reindex(probe_map_df.id)\n",
    "#     .rename(index=ensembl_to_entrez)\n",
    "#     .rename(index=old_to_new_entrez)\n",
    "#     .groupby(level=0).mean()\n",
    "#     .transpose()\n",
    "#     .sort_index(axis='rows')\n",
    "#     .sort_index(axis='columns')\n",
    "# )\n",
    "\n",
    "expr_df.index.rename('sample_id', inplace=True)\n",
    "\n",
    "print(expr_df.shape)\n",
    "expr_df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stratify Balanced Training and Testing Sets in TARGET Gene Expression\n",
    "\n",
    "Output training and testing gene expression datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#strat = pheno_df.set_index('sample_id').reindex(expr_df.index).primary_disease_code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cancertype_count_df = (\n",
    "#     pd.DataFrame(strat.value_counts()) #not using value_counts, what number should we use here? \n",
    "#     .reset_index()\n",
    "#     .rename({'index': 'cancertype', 'primary_disease_code': 'n ='}, axis='columns')\n",
    "# )\n",
    "\n",
    "# file = os.path.join('data', 'target_sample_counts.tsv') #change which file - do we have a file that works for this?\n",
    "# cancertype_count_df.to_csv(file, sep='\\t', index=False)\n",
    "\n",
    "# cancertype_count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df, test_df = train_test_split(expr_df,\n",
    "                                     test_size=0.1,\n",
    "                                     random_state=123) #if no stratify defined, should just randomize on its own"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(16353, 39)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(1818, 39)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(train_df.shape)\n",
    "test_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save train dataframe to file \n",
    "train_file = os.path.join('data', 'train_grein_rat_expression_matrix_processed.tsv.gz')\n",
    "train_df.to_csv(train_file, sep='\\t', compression='gzip', float_format='%.3g')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save test dataframe to file \n",
    "test_file = os.path.join('data', 'test_grein_rat_expression_matrix_processed.tsv.gz')\n",
    "test_df.to_csv(test_file, sep='\\t', compression='gzip', float_format='%.3g')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sort genes based on median absolute deviation and output to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to calculate median absolute deviation\n",
    "def mad(df): \n",
    "    \"\"\"Function to calculate median absolute deviation for a dataframe\n",
    "  argument1 (dataframe): Dataframe for which to calculate the median absolute deviation, row by row\n",
    "\n",
    "  Returns: The median absolute deviation for each row of the data frame\n",
    "  \"\"\"\n",
    "    row_medians = df.median(axis='columns', numeric_only=True)\n",
    "    abs_row_median_diffs = abs(df.sub(row_medians, axis='rows'))\n",
    "    return abs_row_median_diffs.median(axis='columns', numeric_only=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>GSM2668015</th>\n",
       "      <th>GSM2668016</th>\n",
       "      <th>GSM2668017</th>\n",
       "      <th>GSM2668018</th>\n",
       "      <th>GSM2668019</th>\n",
       "      <th>GSM2668020</th>\n",
       "      <th>GSM2668021</th>\n",
       "      <th>GSM2668022</th>\n",
       "      <th>GSM2674552</th>\n",
       "      <th>GSM2674553</th>\n",
       "      <th>GSM2674554</th>\n",
       "      <th>GSM2674555</th>\n",
       "      <th>GSM2674556</th>\n",
       "      <th>GSM2674557</th>\n",
       "      <th>GSM2674558</th>\n",
       "      <th>GSM2674559</th>\n",
       "      <th>GSM2674560</th>\n",
       "      <th>GSM2674561</th>\n",
       "      <th>GSM2674562</th>\n",
       "      <th>GSM2674563</th>\n",
       "      <th>GSM2674564</th>\n",
       "      <th>GSM2674565</th>\n",
       "      <th>GSM2674566</th>\n",
       "      <th>GSM2674567</th>\n",
       "      <th>GSM2683306</th>\n",
       "      <th>GSM2683307</th>\n",
       "      <th>GSM2683308</th>\n",
       "      <th>GSM2683309</th>\n",
       "      <th>GSM2683310</th>\n",
       "      <th>GSM2683311</th>\n",
       "      <th>GSM2683312</th>\n",
       "      <th>GSM2699989</th>\n",
       "      <th>GSM2699990</th>\n",
       "      <th>GSM2699991</th>\n",
       "      <th>GSM2699992</th>\n",
       "      <th>GSM2699993</th>\n",
       "      <th>GSM2699994</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3584</th>\n",
       "      <td>ENSRNOG00000007336Churc1</td>\n",
       "      <td>ENSRNOG00000007336</td>\n",
       "      <td>Churc1</td>\n",
       "      <td>322.7576</td>\n",
       "      <td>237.0001</td>\n",
       "      <td>250.0003</td>\n",
       "      <td>186.1641</td>\n",
       "      <td>221.0005</td>\n",
       "      <td>317.0003</td>\n",
       "      <td>250.6698</td>\n",
       "      <td>235.0003</td>\n",
       "      <td>126.0941</td>\n",
       "      <td>235.1763</td>\n",
       "      <td>355.2647</td>\n",
       "      <td>479.3631</td>\n",
       "      <td>488.3686</td>\n",
       "      <td>465.3487</td>\n",
       "      <td>473.3577</td>\n",
       "      <td>378.5225</td>\n",
       "      <td>187.7590</td>\n",
       "      <td>185.1417</td>\n",
       "      <td>451.4255</td>\n",
       "      <td>441.3341</td>\n",
       "      <td>393.5051</td>\n",
       "      <td>326.2483</td>\n",
       "      <td>336.5778</td>\n",
       "      <td>414.4717</td>\n",
       "      <td>1801.3628</td>\n",
       "      <td>1435.1920</td>\n",
       "      <td>1473.0381</td>\n",
       "      <td>1613.7095</td>\n",
       "      <td>1278.1563</td>\n",
       "      <td>1707.6946</td>\n",
       "      <td>1893.1650</td>\n",
       "      <td>421.6607</td>\n",
       "      <td>433.8507</td>\n",
       "      <td>624.0168</td>\n",
       "      <td>656.0352</td>\n",
       "      <td>693.1579</td>\n",
       "      <td>555.9830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14684</th>\n",
       "      <td>ENSRNOG00000035610Mir377</td>\n",
       "      <td>ENSRNOG00000035610</td>\n",
       "      <td>Mir377</td>\n",
       "      <td>14.0000</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>5.0000</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>6.0000</td>\n",
       "      <td>9.0000</td>\n",
       "      <td>3.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2.9989</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2.0002</td>\n",
       "      <td>5.9990</td>\n",
       "      <td>2.0002</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17130</th>\n",
       "      <td>ENSRNOG00000048445LOC100909862</td>\n",
       "      <td>ENSRNOG00000048445</td>\n",
       "      <td>LOC100909862</td>\n",
       "      <td>8.4862</td>\n",
       "      <td>11.3334</td>\n",
       "      <td>8.6440</td>\n",
       "      <td>9.1396</td>\n",
       "      <td>8.1458</td>\n",
       "      <td>14.6378</td>\n",
       "      <td>13.0723</td>\n",
       "      <td>6.8770</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>13.7399</td>\n",
       "      <td>4.4305</td>\n",
       "      <td>11.3453</td>\n",
       "      <td>0.0081</td>\n",
       "      <td>7.7333</td>\n",
       "      <td>7.0126</td>\n",
       "      <td>8.2752</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>ENSRNOG00000000169Spata5l1</td>\n",
       "      <td>ENSRNOG00000000169</td>\n",
       "      <td>Spata5l1</td>\n",
       "      <td>225.5967</td>\n",
       "      <td>150.5534</td>\n",
       "      <td>161.6999</td>\n",
       "      <td>129.2117</td>\n",
       "      <td>132.6296</td>\n",
       "      <td>188.6234</td>\n",
       "      <td>143.0324</td>\n",
       "      <td>132.2708</td>\n",
       "      <td>116.8599</td>\n",
       "      <td>102.3285</td>\n",
       "      <td>195.1699</td>\n",
       "      <td>179.9838</td>\n",
       "      <td>169.8034</td>\n",
       "      <td>161.4641</td>\n",
       "      <td>175.8895</td>\n",
       "      <td>186.4348</td>\n",
       "      <td>78.1931</td>\n",
       "      <td>88.6635</td>\n",
       "      <td>108.4148</td>\n",
       "      <td>124.4535</td>\n",
       "      <td>131.3203</td>\n",
       "      <td>98.6024</td>\n",
       "      <td>117.3997</td>\n",
       "      <td>129.5221</td>\n",
       "      <td>405.9823</td>\n",
       "      <td>348.4350</td>\n",
       "      <td>245.8922</td>\n",
       "      <td>376.0369</td>\n",
       "      <td>228.2067</td>\n",
       "      <td>371.6765</td>\n",
       "      <td>287.5426</td>\n",
       "      <td>153.5827</td>\n",
       "      <td>198.6694</td>\n",
       "      <td>239.0798</td>\n",
       "      <td>421.1322</td>\n",
       "      <td>405.5692</td>\n",
       "      <td>353.1912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17164</th>\n",
       "      <td>ENSRNOG00000048549Olr1240</td>\n",
       "      <td>ENSRNOG00000048549</td>\n",
       "      <td>Olr1240</td>\n",
       "      <td>31.3617</td>\n",
       "      <td>16.1621</td>\n",
       "      <td>14.9487</td>\n",
       "      <td>12.3349</td>\n",
       "      <td>11.9975</td>\n",
       "      <td>22.6926</td>\n",
       "      <td>24.3146</td>\n",
       "      <td>15.8916</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>19.7699</td>\n",
       "      <td>12.4066</td>\n",
       "      <td>13.0108</td>\n",
       "      <td>20.8329</td>\n",
       "      <td>10.6476</td>\n",
       "      <td>20.3194</td>\n",
       "      <td>18.4805</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  gene_id                gene   gene_symbol  \\\n",
       "sample_id                                                                     \n",
       "3584             ENSRNOG00000007336Churc1  ENSRNOG00000007336        Churc1   \n",
       "14684            ENSRNOG00000035610Mir377  ENSRNOG00000035610        Mir377   \n",
       "17130      ENSRNOG00000048445LOC100909862  ENSRNOG00000048445  LOC100909862   \n",
       "66             ENSRNOG00000000169Spata5l1  ENSRNOG00000000169      Spata5l1   \n",
       "17164           ENSRNOG00000048549Olr1240  ENSRNOG00000048549       Olr1240   \n",
       "\n",
       "           GSM2668015  GSM2668016  GSM2668017  GSM2668018  GSM2668019  \\\n",
       "sample_id                                                               \n",
       "3584         322.7576    237.0001    250.0003    186.1641    221.0005   \n",
       "14684         14.0000      7.0000      5.0000      8.0000      7.0000   \n",
       "17130          8.4862     11.3334      8.6440      9.1396      8.1458   \n",
       "66           225.5967    150.5534    161.6999    129.2117    132.6296   \n",
       "17164         31.3617     16.1621     14.9487     12.3349     11.9975   \n",
       "\n",
       "           GSM2668020  GSM2668021  GSM2668022  GSM2674552  GSM2674553  \\\n",
       "sample_id                                                               \n",
       "3584         317.0003    250.6698    235.0003    126.0941    235.1763   \n",
       "14684          6.0000      9.0000      3.0000      0.0000      0.0000   \n",
       "17130         14.6378     13.0723      6.8770      0.0000      0.0000   \n",
       "66           188.6234    143.0324    132.2708    116.8599    102.3285   \n",
       "17164         22.6926     24.3146     15.8916      0.0000      0.0000   \n",
       "\n",
       "           GSM2674554  GSM2674555  GSM2674556  GSM2674557  GSM2674558  \\\n",
       "sample_id                                                               \n",
       "3584         355.2647    479.3631    488.3686    465.3487    473.3577   \n",
       "14684          0.0000      0.0000      0.0000      0.0000      0.0000   \n",
       "17130          0.0000      0.0000      0.0000      0.0000      0.0000   \n",
       "66           195.1699    179.9838    169.8034    161.4641    175.8895   \n",
       "17164          0.0000      0.0000      0.0000      0.0000      0.0000   \n",
       "\n",
       "           GSM2674559  GSM2674560  GSM2674561  GSM2674562  GSM2674563  \\\n",
       "sample_id                                                               \n",
       "3584         378.5225    187.7590    185.1417    451.4255    441.3341   \n",
       "14684          0.0000      0.0000      0.0000      0.0000      0.0000   \n",
       "17130          0.0000      0.0000      0.0000      0.0000      0.0000   \n",
       "66           186.4348     78.1931     88.6635    108.4148    124.4535   \n",
       "17164          0.0000      0.0000      0.0000      0.0000      0.0000   \n",
       "\n",
       "           GSM2674564  GSM2674565  GSM2674566  GSM2674567  GSM2683306  \\\n",
       "sample_id                                                               \n",
       "3584         393.5051    326.2483    336.5778    414.4717   1801.3628   \n",
       "14684          0.0000      0.0000      0.0000      0.0000      2.9989   \n",
       "17130          0.0000      0.0000      0.0000      0.0000     13.7399   \n",
       "66           131.3203     98.6024    117.3997    129.5221    405.9823   \n",
       "17164          0.0000      0.0000      0.0000      0.0000     19.7699   \n",
       "\n",
       "           GSM2683307  GSM2683308  GSM2683309  GSM2683310  GSM2683311  \\\n",
       "sample_id                                                               \n",
       "3584        1435.1920   1473.0381   1613.7095   1278.1563   1707.6946   \n",
       "14684          0.0000      0.0000      0.0000      2.0002      5.9990   \n",
       "17130          4.4305     11.3453      0.0081      7.7333      7.0126   \n",
       "66           348.4350    245.8922    376.0369    228.2067    371.6765   \n",
       "17164         12.4066     13.0108     20.8329     10.6476     20.3194   \n",
       "\n",
       "           GSM2683312  GSM2699989  GSM2699990  GSM2699991  GSM2699992  \\\n",
       "sample_id                                                               \n",
       "3584        1893.1650    421.6607    433.8507    624.0168    656.0352   \n",
       "14684          2.0002      0.0000      0.0000      0.0000      0.0000   \n",
       "17130          8.2752      0.0000      0.0000      0.0000      0.0000   \n",
       "66           287.5426    153.5827    198.6694    239.0798    421.1322   \n",
       "17164         18.4805      0.0000      0.0000      0.0000      0.0000   \n",
       "\n",
       "           GSM2699993  GSM2699994  \n",
       "sample_id                          \n",
       "3584         693.1579    555.9830  \n",
       "14684          0.0000      0.0000  \n",
       "17130          0.0000      0.0000  \n",
       "66           405.5692    353.1912  \n",
       "17164          0.0000      0.0000  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add gene_id as a column in dataframe\n",
    "train_df['gene_id'] = train_df['gene'] + train_df['gene_symbol']\n",
    "col = train_df.pop('gene_id')\n",
    "train_df.insert(0, col.name, col)\n",
    "\n",
    "train_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample_id\n",
       "14526    104386.5201\n",
       "13990     98591.7197\n",
       "13670     71535.2136\n",
       "13750     68463.5045\n",
       "13940     65865.1337\n",
       "            ...     \n",
       "17213         0.0000\n",
       "16311         0.0000\n",
       "17243         0.0000\n",
       "8596          0.0000\n",
       "9177          0.0000\n",
       "Length: 16353, dtype: float64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Determine most variably expressed genes and subset\n",
    "# create dataframe to save median absolute deviation data for rats\n",
    "train_df_mad = mad(train_df.drop(['gene_id','gene', 'gene_symbol'], axis = 1))\n",
    "train_df_mad.columns = ['gene_id', 'median_abs_deviation']\n",
    "\n",
    "train_df_mad = train_df_mad.sort_values(ascending=False)\n",
    "train_df_mad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to tsv file\n",
    "file = os.path.join('data', 'grein_mad_rat_genes.tsv')\n",
    "train_df_mad.to_csv(file, sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
